// @taiga-ui/cdk/esm2022/directives/transitioned/transitioned.directive.mjs
import { Directive, inject, NgZone } from '@angular/core';
import { tuiInjectElement } from '@taiga-ui/cdk/utils';
import * as i0 from "@angular/core";
class TuiTransitioned {
    constructor() {
        const el = tuiInjectElement();
        inject(NgZone).runOutsideAngular(() => {
            setTimeout(() => {
                el.style.transition = '';
            });
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTransitioned, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTransitioned, isStandalone: true, selector: "[tuiTransitioned]", host: { properties: { "style.transition": "\"none\"" } }, ngImport: i0 }); }
}
export { TuiTransitioned };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTransitioned, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiTransitioned]',
                    host: {
                        '[style.transition]': '"none"',
                    },
                }]
        }], ctorParameters: function () { return []; } });
