// @taiga-ui/layout/esm2022/components/dynamic-header/dynamic-header-container.directive.mjs
import { ContentChildren, Directive, signal, } from '@angular/core';
import { toSignal } from '@angular/core/rxjs-interop';
import { WaIntersectionObserverDirective, WaIntersectionRoot, } from '@ng-web-apis/intersection-observer';
import { EMPTY_QUERY } from '@taiga-ui/cdk/constants';
import { tuiZoneOptimized } from '@taiga-ui/cdk/observables';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { distinctUntilChanged, fromEvent, map, pairwise } from 'rxjs';
import { TuiDynamicHeaderAnchorDirective } from './dynamic-header-anchor.directive';
import * as i0 from "@angular/core";
import * as i1 from "@ng-web-apis/intersection-observer";
class TuiDynamicHeaderContainerDirective {
    constructor() {
        this.el = tuiInjectElement();
        this.headers = EMPTY_QUERY;
        this.hiddenHeaders = signal([]);
        this.scrollDir = toSignal(fromEvent(this.el, 'scroll').pipe(map(({ target }) => target.scrollTop), pairwise(), map(([prev, next]) => (next > prev ? -1 : 1)), distinctUntilChanged(), tuiZoneOptimized()));
    }
    update() {
        this.hiddenHeaders.set(this.headers.filter((h) => h.visible()).map((h) => h.templateRef));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDynamicHeaderContainerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiDynamicHeaderContainerDirective, isStandalone: true, selector: "[tuiDynamicHeaderContainer]", host: { attributes: { "waIntersectionRootMargin": "-44px 0px 1000000% 0px" } }, queries: [{ propertyName: "headers", predicate: TuiDynamicHeaderAnchorDirective, descendants: true }], hostDirectives: [{ directive: i1.WaIntersectionRoot }, { directive: i1.WaIntersectionObserverDirective }], ngImport: i0 }); }
}
export { TuiDynamicHeaderContainerDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDynamicHeaderContainerDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiDynamicHeaderContainer]',
                    hostDirectives: [WaIntersectionRoot, WaIntersectionObserverDirective],
                    host: {
                        waIntersectionRootMargin: '-44px 0px 1000000% 0px',
                    },
                }]
        }], propDecorators: { headers: [{
                type: ContentChildren,
                args: [TuiDynamicHeaderAnchorDirective, {
                        descendants: true,
                    }]
            }] } });
