// @taiga-ui/layout/esm2022/components/dynamic-header/dynamic-header-anchor.directive.mjs
import { Directive, inject, signal, TemplateRef, ViewContainerRef, } from '@angular/core';
import { WaIntersectionObserverDirective } from '@ng-web-apis/intersection-observer';
import { TuiDynamicHeaderContainerDirective } from './dynamic-header-container.directive';
import * as i0 from "@angular/core";
class TuiDynamicHeaderAnchorDirective {
    constructor() {
        this.viewContainer = inject(ViewContainerRef);
        this.observer = inject(WaIntersectionObserverDirective);
        this.container = inject(TuiDynamicHeaderContainerDirective);
        this.templateRef = inject((TemplateRef));
        this.view = this.viewContainer.createEmbeddedView(this.templateRef);
        this.visible = signal(false);
    }
    ngAfterViewInit() {
        this.observer.observe(this.view.rootNodes[0], ([e]) => {
            this.visible.set(!e?.isIntersecting);
            this.container.update();
        });
    }
    ngOnDestroy() {
        this.observer.unobserve(this.view.rootNodes[0]);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDynamicHeaderAnchorDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiDynamicHeaderAnchorDirective, isStandalone: true, selector: "[tuiDynamicHeaderAnchor]", ngImport: i0 }); }
}
export { TuiDynamicHeaderAnchorDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDynamicHeaderAnchorDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiDynamicHeaderAnchor]',
                }]
        }] });
