// @taiga-ui/kit/esm2022/directives/copy/button-copy.component.mjs
import { Clipboard } from '@angular/cdk/clipboard';
import { ChangeDetectionStrategy, Component, computed, inject, Input, signal, ViewEncapsulation, } from '@angular/core';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { tuiDirectiveBinding, tuiIsString } from '@taiga-ui/cdk/utils/miscellaneous';
import { TuiButton, tuiButtonOptionsProvider } from '@taiga-ui/core/components/button';
import { TUI_NOTIFICATION_OPTIONS } from '@taiga-ui/core/components/notification';
import { TuiHintDirective } from '@taiga-ui/core/directives/hint';
import { TUI_DARK_MODE } from '@taiga-ui/core/tokens';
import { TuiIcons } from '@taiga-ui/core/directives/icons';
import { TUI_COPY_TEXTS } from '@taiga-ui/kit/tokens';
import { toSignal } from '@angular/core/rxjs-interop';
import { NgIf } from '@angular/common';
import { TUI_COPY_OPTIONS } from './copy.options';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/components/button";
class TuiButtonCopyComponent {
    constructor() {
        this.el = tuiInjectElement();
        this.clipboard = inject(Clipboard);
        this.options = inject(TUI_COPY_OPTIONS);
        this.copyTexts = toSignal(inject(TUI_COPY_TEXTS));
        this.notification = inject(TUI_NOTIFICATION_OPTIONS);
        this.hint = inject(TuiHintDirective);
        this.check = tuiIsString(this.notification.icon)
            ? this.notification.icon
            : this.notification.icon('positive');
        this.appearance = this.hint.appearance ||
            this.hint.el.closest('[tuiTheme]')?.getAttribute('tuiTheme');
        this.darkMode = inject(TUI_DARK_MODE);
        this.copied = signal(false);
        this.copiedText = computed(() => this.copyTexts()?.[1] ?? '');
        this.icon = tuiDirectiveBinding(TuiIcons, 'iconStart', computed(() => {
            if (this.copied()) {
                return this.check;
            }
            return typeof this.options.icon === 'string'
                ? this.options.icon
                : this.options.icon('s');
        }));
        this.tuiButtonCopy = '';
    }
    copy(target) {
        const copied = this.el === target;
        if (copied) {
            this.clipboard.copy(this.tuiButtonCopy);
        }
        this.copied.set(copied);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiButtonCopyComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiButtonCopyComponent, isStandalone: true, selector: "[tuiButtonCopy]", inputs: { tuiButtonCopy: "tuiButtonCopy" }, host: { attributes: { "tuiButton": "", "type": "button" }, listeners: { "document:click.capture": "copy($event.target)" }, properties: { "attr.tuiTheme": "appearance === \"dark\" ? \"light\" : darkMode() ? \"light\" : \"dark\"" } }, providers: [tuiButtonOptionsProvider({ appearance: 'flat-grayscale', size: 's' })], hostDirectives: [{ directive: i1.TuiButton }], ngImport: i0, template: `
        <ng-container *ngIf="copied(); else notCopied">
            {{ copiedText() }}
        </ng-container>
        <ng-template #notCopied>
            <ng-content />
        </ng-template>
    `, isInline: true, styles: ["tui-hint:has([tuiButtonCopy]){display:flex;padding:.375rem;max-inline-size:none}[tuiButtonCopy]:not(:first-of-type){margin-inline-start:calc(.75rem - 1px);border-image:linear-gradient(transparent .375rem,var(--tui-border-normal) .375rem,var(--tui-border-normal) calc(100% - .375rem),transparent calc(100% - .375rem)) 0 0 0 1 / 0 1px / 0 .375rem}\n"], dependencies: [{ kind: "directive", type: NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
export { TuiButtonCopyComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiButtonCopyComponent, decorators: [{
            type: Component,
            args: [{ standalone: true, selector: '[tuiButtonCopy]', imports: [NgIf], template: `
        <ng-container *ngIf="copied(); else notCopied">
            {{ copiedText() }}
        </ng-container>
        <ng-template #notCopied>
            <ng-content />
        </ng-template>
    `, encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, providers: [tuiButtonOptionsProvider({ appearance: 'flat-grayscale', size: 's' })], hostDirectives: [TuiButton], host: {
                        tuiButton: '',
                        type: 'button',
                        '[attr.tuiTheme]': 'appearance === "dark" ? "light" : darkMode() ? "light" : "dark"',
                        '(document:click.capture)': 'copy($event.target)',
                    }, styles: ["tui-hint:has([tuiButtonCopy]){display:flex;padding:.375rem;max-inline-size:none}[tuiButtonCopy]:not(:first-of-type){margin-inline-start:calc(.75rem - 1px);border-image:linear-gradient(transparent .375rem,var(--tui-border-normal) .375rem,var(--tui-border-normal) calc(100% - .375rem),transparent calc(100% - .375rem)) 0 0 0 1 / 0 1px / 0 .375rem}\n"] }]
        }], propDecorators: { tuiButtonCopy: [{
                type: Input
            }] } });
