// @taiga-ui/kit/esm2022/tokens/items-handlers.mjs
import { InjectionToken } from '@angular/core';
import { TUI_DEFAULT_IDENTITY_MATCHER, TUI_FALSE_HANDLER } from '@taiga-ui/cdk/constants';
import { tuiProvideOptions } from '@taiga-ui/cdk/utils/miscellaneous';
/**
 * @deprecated use it only for LEGACY controls. For new controls use the same entity from `@taiga-ui/core`.
 * TODO(v5): delete
 */
export const TUI_DEFAULT_ITEMS_HANDLERS = {
    stringify: String,
    identityMatcher: TUI_DEFAULT_IDENTITY_MATCHER,
    disabledItemHandler: TUI_FALSE_HANDLER,
};
/**
 * @deprecated use it only for LEGACY controls. For new controls use the same entity from `@taiga-ui/core`.
 * TODO(v5): delete
 */
export const TUI_ITEMS_HANDLERS = new InjectionToken(ngDevMode ? 'TUI_ITEMS_HANDLERS' : '', {
    factory: () => TUI_DEFAULT_ITEMS_HANDLERS,
});
/**
 * @deprecated use it only for LEGACY controls. For new controls use the same entity from `@taiga-ui/core`.
 * TODO(v5): delete
 */
export function tuiItemsHandlersProvider(options) {
    return tuiProvideOptions(TUI_ITEMS_HANDLERS, options, TUI_DEFAULT_ITEMS_HANDLERS);
}
