// @taiga-ui/cdk/esm2022/directives/droppable/droppable.directive.mjs
import { Directive, Output } from '@angular/core';
import { tuiPreventDefault, tuiTypedFromEvent } from '@taiga-ui/cdk/observables';
import { tuiInjectElement } from '@taiga-ui/cdk/utils';
import { tuiIsPresent } from '@taiga-ui/cdk/utils/miscellaneous';
import { distinctUntilChanged, filter, map, merge, startWith, switchMap } from 'rxjs';
import * as i0 from "@angular/core";
class TuiDroppable {
    constructor() {
        this.el = tuiInjectElement();
        this.tuiDroppableDropped = tuiTypedFromEvent(this.el, 'drop').pipe(tuiPreventDefault(), map((event) => event.dataTransfer), filter(tuiIsPresent));
        this.tuiDroppableDragOverChange = tuiTypedFromEvent(this.el, 'dragenter').pipe(switchMap(({ target, dataTransfer }) => merge(tuiTypedFromEvent(this.el, 'dragleave').pipe(filter((event) => event.target === target)), tuiTypedFromEvent(this.el, 'drop')).pipe(map(() => null), startWith(dataTransfer))), distinctUntilChanged((a, b) => (!!a && !!b) || (!a && !b)));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDroppable, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiDroppable, isStandalone: true, selector: "[tuiDroppableDropped], [tuiDroppableDragOverChange]", outputs: { tuiDroppableDropped: "tuiDroppableDropped", tuiDroppableDragOverChange: "tuiDroppableDragOverChange" }, host: { listeners: { "dragover.prevent.zoneless": "0" } }, ngImport: i0 }); }
}
export { TuiDroppable };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDroppable, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiDroppableDropped], [tuiDroppableDragOverChange]',
                    host: {
                        '(dragover.prevent.zoneless)': '0',
                    },
                }]
        }], propDecorators: { tuiDroppableDropped: [{
                type: Output
            }], tuiDroppableDragOverChange: [{
                type: Output
            }] } });
