// @taiga-ui/kit/esm2022/components/input-time/input-time.component.mjs
import { NgIf } from '@angular/common';
import { ChangeDetectionStrategy, Component, computed, Directive, inject, ViewEncapsulation, } from '@angular/core';
import { TuiControl } from '@taiga-ui/cdk/classes';
import { TuiTime } from '@taiga-ui/cdk/date-time';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { TuiTextfieldContent, TuiTextfieldDirective, TuiWithNativePicker, } from '@taiga-ui/core/components/textfield';
import { TuiInputTimeDirective } from './input-time.directive';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/components/textfield";
class TuiNativeTimePicker {
    constructor() {
        this.list = tuiInjectElement().getAttribute('list');
    }
    getStep(timeMode) {
        switch (timeMode) {
            case 'HH:MM:SS':
            case 'HH:MM:SS AA':
                return 1;
            case 'HH:MM:SS.MSS':
            case 'HH:MM:SS.MSS AA':
                return 0.001;
            default:
                return 60;
        }
    }
    toISOString(value) {
        const [day, time] = Array.isArray(value) ? value : [null, value];
        const dateString = day ? day.toJSON() + (time ? 'T' : '') : '';
        const timeString = time ? time.toString('HH:MM:SS.MSS') : '';
        return dateString + timeString;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiNativeTimePicker, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiNativeTimePicker, host: { properties: { "attr.list": "null" } }, ngImport: i0 }); }
}
export { TuiNativeTimePicker };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiNativeTimePicker, decorators: [{
            type: Directive,
            args: [{
                    host: {
                        '[attr.list]': 'null',
                    },
                }]
        }] });
class TuiInputTimeComponent extends TuiNativeTimePicker {
    constructor() {
        super(...arguments);
        this.control = inject(TuiControl);
        this.host = inject(TuiInputTimeDirective);
        this.textfield = inject(TuiTextfieldDirective);
        this.value = computed(() => this.toISOString(this.control.value()));
        this.step = computed(() => this.getStep(this.host.timeMode()));
    }
    setValue(value) {
        this.host.setValue(TuiTime.fromString(value));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputTimeComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiInputTimeComponent, isStandalone: true, selector: "input[tuiInputTime][type=\"time\"]", host: { attributes: { "ngSkipHydration": "true" } }, usesInheritance: true, hostDirectives: [{ directive: i1.TuiWithNativePicker }], ngImport: i0, template: "<ng-container *ngIf=\"host.native\">\n    <input\n        *tuiTextfieldContent\n        type=\"time\"\n        [attr.list]=\"list\"\n        [step]=\"step()\"\n        [value]=\"value()\"\n        (change)=\"setValue($any($event.target).value)\"\n    />\n</ng-container>\n", styles: ["tui-textfield input[tuiInputTime]~.t-content input[type=time]{position:absolute;right:0;left:auto;inline-size:calc(var(--t-right) + var(--t-padding));opacity:0;margin:0;padding:0}tui-textfield input[tuiInputTime]~.t-content input[type=time]::-webkit-calendar-picker-indicator{position:absolute;top:0;left:0;inline-size:100%;block-size:100%}\n"], dependencies: [{ kind: "directive", type: NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: TuiTextfieldContent, selector: "ng-template[tuiTextfieldContent]" }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
export { TuiInputTimeComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputTimeComponent, decorators: [{
            type: Component,
            args: [{ standalone: true, selector: 'input[tuiInputTime][type="time"]', imports: [NgIf, TuiTextfieldContent], encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, hostDirectives: [TuiWithNativePicker], host: { ngSkipHydration: 'true' }, template: "<ng-container *ngIf=\"host.native\">\n    <input\n        *tuiTextfieldContent\n        type=\"time\"\n        [attr.list]=\"list\"\n        [step]=\"step()\"\n        [value]=\"value()\"\n        (change)=\"setValue($any($event.target).value)\"\n    />\n</ng-container>\n", styles: ["tui-textfield input[tuiInputTime]~.t-content input[type=time]{position:absolute;right:0;left:auto;inline-size:calc(var(--t-right) + var(--t-padding));opacity:0;margin:0;padding:0}tui-textfield input[tuiInputTime]~.t-content input[type=time]::-webkit-calendar-picker-indicator{position:absolute;top:0;left:0;inline-size:100%;block-size:100%}\n"] }]
        }] });
