// @taiga-ui/font-watcher/index.esm.js
/* eslint-disable @typescript-eslint/no-unnecessary-condition, compat/compat */
const IFRAME = {
    position: 'fixed',
    visibility: 'hidden',
    pointerEvents: 'none',
};
const BODY = {
    height: 'fit-content',
    lineHeight: '1em',
    fontSize: 'calc(env(preferred-text-scale) * 1em)',
};
function tuiFontSizeWatcher(callback, iframe = globalThis.document.createElement('iframe')) {
    const resize = () => {
        const { innerWidth = 0, outerWidth = 0, devicePixelRatio = 0, } = iframe.ownerDocument?.defaultView || {};
        iframe.width = `${innerWidth === outerWidth ? innerWidth : innerWidth / devicePixelRatio}`;
    };
    iframe.ownerDocument?.body.append(iframe);
    iframe.ownerDocument?.defaultView?.addEventListener('resize', resize);
    const doc = iframe.contentDocument;
    const observer = new ResizeObserver(() => callback(doc?.body.offsetHeight || 0));
    Object.assign(iframe.style, IFRAME);
    Object.assign(doc?.body.style || {}, BODY);
    doc?.documentElement.style.setProperty('font', '-apple-system-body');
    doc?.body.insertAdjacentText('beforeend', '.'.repeat(1000));
    observer.observe(doc?.body || iframe);
    resize();
    return () => {
        observer.disconnect();
        iframe.ownerDocument?.defaultView?.removeEventListener('resize', resize);
        iframe.remove();
    };
}

export { tuiFontSizeWatcher };
