// @taiga-ui/kit/esm2022/components/multi-select/multi-select-group/multi-select-group.directive.mjs
import { Directive } from '@angular/core';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { tuiAsOptionContent } from '@taiga-ui/core/components/data-list';
import { TuiMultiSelectOption } from '../multi-select-option/multi-select-option.component';
import * as i0 from "@angular/core";
class TuiMultiSelectGroupDirective {
    constructor() {
        tuiInjectElement()
            .closest('[role="listbox"]')
            ?.setAttribute('aria-multiselectable', 'true');
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectGroupDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiMultiSelectGroupDirective, isStandalone: true, selector: "[tuiMultiSelectGroup]", providers: [tuiAsOptionContent(TuiMultiSelectOption)], ngImport: i0 }); }
}
export { TuiMultiSelectGroupDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectGroupDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiMultiSelectGroup]',
                    providers: [tuiAsOptionContent(TuiMultiSelectOption)],
                }]
        }], ctorParameters: function () { return []; } });
