// @taiga-ui/core/esm2022/directives/items-handlers/items-handlers.tokens.mjs
import { InjectionToken, Optional, signal, SkipSelf, } from '@angular/core';
import { TUI_DEFAULT_IDENTITY_MATCHER, TUI_FALSE_HANDLER } from '@taiga-ui/cdk/constants';
export const TUI_DEFAULT_ITEMS_HANDLERS = {
    stringify: signal(String),
    identityMatcher: signal(TUI_DEFAULT_IDENTITY_MATCHER),
    disabledItemHandler: signal(TUI_FALSE_HANDLER),
};
/**
 * Default items handlers for components
 */
export const TUI_ITEMS_HANDLERS = new InjectionToken(ngDevMode ? 'TUI_ITEMS_HANDLERS' : '', {
    factory: () => TUI_DEFAULT_ITEMS_HANDLERS,
});
export function tuiItemsHandlersProvider(options) {
    return {
        provide: TUI_ITEMS_HANDLERS,
        deps: [[new Optional(), new SkipSelf(), TUI_ITEMS_HANDLERS]],
        useFactory: (parent) => ({
            stringify: signal(parent?.stringify() ?? TUI_DEFAULT_ITEMS_HANDLERS.stringify()),
            identityMatcher: signal(parent?.identityMatcher() ?? TUI_DEFAULT_ITEMS_HANDLERS.identityMatcher()),
            disabledItemHandler: signal(parent?.disabledItemHandler() ??
                TUI_DEFAULT_ITEMS_HANDLERS.disabledItemHandler()),
            ...options,
        }),
    };
}
