// @maskito/angular/esm2022/lib/maskito.directive.mjs
import { Directive, effect, ElementRef, inject, Input, NgZone, signal, untracked, } from '@angular/core';
import { DefaultValueAccessor } from '@angular/forms';
import { Maskito, MASKITO_DEFAULT_ELEMENT_PREDICATE, maskitoTransform, } from '@maskito/core';
import * as i0 from "@angular/core";
class MaskitoDirective {
    constructor() {
        this.elementRef = inject(ElementRef).nativeElement;
        this.ngZone = inject(NgZone);
        this.maskedElement = null;
        // TODO(v4): delete excessive setters/getters
        this.optionsInternal = signal(null);
        this.elementPredicateInternal = signal(MASKITO_DEFAULT_ELEMENT_PREDICATE);
        this.initEffect = effect(async () => {
            const elementPredicate = this.elementPredicateInternal();
            const options = this.optionsInternal();
            const { maskedElement, elementRef, ngZone } = this;
            maskedElement?.destroy();
            if (!options) {
                return;
            }
            const predicateResult = await elementPredicate(elementRef);
            if (this.elementPredicate !== elementPredicate || this.options !== options) {
                // Ignore the result of the predicate if the
                // maskito element (or its options) has changed before the predicate was resolved.
                return;
            }
            ngZone.runOutsideAngular(() => {
                this.maskedElement = new Maskito(predicateResult, options);
            });
        });
        const accessor = inject(DefaultValueAccessor, { self: true, optional: true });
        if (accessor) {
            const original = accessor.writeValue.bind(accessor);
            accessor.writeValue = (value) => {
                const options = untracked(this.optionsInternal);
                original(options ? maskitoTransform(String(value ?? ''), options) : value);
            };
        }
    }
    set options(value) {
        this.optionsInternal.set(value);
    }
    set elementPredicate(value) {
        this.elementPredicateInternal.set(value);
    }
    get options() {
        return this.optionsInternal();
    }
    get elementPredicate() {
        return this.elementPredicateInternal();
    }
    // eslint-disable-next-line @angular-eslint/no-empty-lifecycle-method
    async ngOnChanges() {
        // TODO(v4): delete
    }
    ngOnDestroy() {
        this.maskedElement?.destroy();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: MaskitoDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: MaskitoDirective, isStandalone: true, selector: "[maskito]", inputs: { options: ["maskito", "options"], elementPredicate: ["maskitoElement", "elementPredicate"] }, usesOnChanges: true, ngImport: i0 }); }
}
export { MaskitoDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: MaskitoDirective, decorators: [{
            type: Directive,
            args: [{ standalone: true, selector: '[maskito]' }]
        }], ctorParameters: function () { return []; }, propDecorators: { options: [{
                type: Input,
                args: ['maskito']
            }], elementPredicate: [{
                type: Input,
                args: ['maskitoElement']
            }] } });
