// @maskito/angular/esm2022/lib/pattern.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { MaskitoDirective } from './maskito.directive';
import * as i0 from "@angular/core";
import * as i1 from "./maskito.directive";
class MaskitoPattern {
    constructor() {
        this.maskitoDirective = inject(MaskitoDirective, { self: true });
    }
    set regExpStr(pattern) {
        this.maskitoDirective.options = {
            mask: typeof pattern === 'string' ? new RegExp(`^${pattern}$`) : pattern,
        };
        this.maskitoDirective.ngOnChanges();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: MaskitoPattern, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: MaskitoPattern, isStandalone: true, selector: "[maskitoPattern]", inputs: { regExpStr: ["maskitoPattern", "regExpStr"] }, hostDirectives: [{ directive: i1.MaskitoDirective }], ngImport: i0 }); }
}
export { MaskitoPattern };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: MaskitoPattern, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[maskitoPattern]',
                    hostDirectives: [MaskitoDirective],
                }]
        }], propDecorators: { regExpStr: [{
                type: Input,
                args: ['maskitoPattern']
            }] } });
