// @taiga-ui/core/esm2022/directives/hint/hint-overflow.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { TuiHintDirective } from './hint.directive';
import * as i0 from "@angular/core";
import * as i1 from "./hint.directive";
class TuiHintOverflow {
    constructor() {
        this.hint = inject(TuiHintDirective);
        this.tuiHintOverflow = '';
    }
    onMouseEnter({ scrollWidth, clientWidth, textContent }) {
        this.hint.tuiHint =
            scrollWidth > clientWidth && this.tuiHintOverflow !== null
                ? this.tuiHintOverflow || textContent
                : '';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiHintOverflow, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiHintOverflow, isStandalone: true, selector: "[tuiHintOverflow]", inputs: { tuiHintOverflow: "tuiHintOverflow" }, host: { listeners: { "mouseenter": "onMouseEnter($event.currentTarget)" } }, hostDirectives: [{ directive: i1.TuiHintDirective, inputs: ["tuiHintAppearance", "tuiHintAppearance"] }], ngImport: i0 }); }
}
export { TuiHintOverflow };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiHintOverflow, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiHintOverflow]',
                    hostDirectives: [
                        {
                            directive: TuiHintDirective,
                            inputs: ['tuiHintAppearance'],
                        },
                    ],
                    host: {
                        '(mouseenter)': 'onMouseEnter($event.currentTarget)',
                    },
                }]
        }], propDecorators: { tuiHintOverflow: [{
                type: Input
            }] } });
