// @taiga-ui/kit/fesm2022/taiga-ui-kit-components-slides.mjs
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Directive, Input } from '@angular/core';
import * as i1 from '@taiga-ui/cdk/directives/animated';
import { TuiAnimatedParent } from '@taiga-ui/cdk/directives/animated';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { tuiWithStyles } from '@taiga-ui/cdk/utils/miscellaneous';

class TuiSlidesStyles {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSlidesStyles, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiSlidesStyles, isStandalone: true, selector: "ng-component", host: { classAttribute: "tui-slides" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiSlides]{display:grid;isolation:isolate}[tuiSlides][data-direction=\"-1\"]>*{--tui-from: translateX(-2rem)}[tuiSlides][data-direction=\"-1\"]>*.tui-leave{--tui-from: translateX(1rem)}[tuiSlides][data-direction=\"1\"]>*{--tui-from: translateX(2rem)}[tuiSlides][data-direction=\"1\"]>*.tui-leave{--tui-from: translateX(-1rem)}[tuiSlides]>*{grid-area:1 / 1;align-self:start;mix-blend-mode:plus-lighter;animation-name:tuiFade,tuiSlide;animation-duration:var(--tui-duration);animation-timing-function:ease-in-out;--tui-from: translateY(0)}[tuiSlides]>*.tui-slide:not(.tui-leave){animation-name:none}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSlidesStyles, decorators: [{
            type: Component,
            args: [{ standalone: true, template: '', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        class: 'tui-slides',
                    }, styles: ["[tuiSlides]{display:grid;isolation:isolate}[tuiSlides][data-direction=\"-1\"]>*{--tui-from: translateX(-2rem)}[tuiSlides][data-direction=\"-1\"]>*.tui-leave{--tui-from: translateX(1rem)}[tuiSlides][data-direction=\"1\"]>*{--tui-from: translateX(2rem)}[tuiSlides][data-direction=\"1\"]>*.tui-leave{--tui-from: translateX(-1rem)}[tuiSlides]>*{grid-area:1 / 1;align-self:start;mix-blend-mode:plus-lighter;animation-name:tuiFade,tuiSlide;animation-duration:var(--tui-duration);animation-timing-function:ease-in-out;--tui-from: translateY(0)}[tuiSlides]>*.tui-slide:not(.tui-leave){animation-name:none}\n"] }]
        }] });
class TuiSlides {
    constructor() {
        this.el = tuiInjectElement();
        this.nothing = tuiWithStyles(TuiSlidesStyles);
        this.direction = '';
    }
    get sign() {
        return Math.sign(this.direction || 0);
    }
    onAnimation(target) {
        Array.from(this.el.children)
            .find((element) => element === target)
            ?.classList.add('tui-slide');
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSlides, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiSlides, isStandalone: true, selector: "[tuiSlides]", inputs: { direction: ["tuiSlides", "direction"] }, host: { attributes: { "tuiSlides": "" }, listeners: { "animationend": "onAnimation($event.target)" }, properties: { "attr.data-direction": "sign" } }, hostDirectives: [{ directive: i1.TuiAnimatedParent }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSlides, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiSlides]',
                    hostDirectives: [TuiAnimatedParent],
                    host: {
                        tuiSlides: '',
                        '[attr.data-direction]': 'sign',
                        '(animationend)': 'onAnimation($event.target)',
                    },
                }]
        }], propDecorators: { direction: [{
                type: Input,
                args: ['tuiSlides']
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiSlides };
