// @taiga-ui/kit/esm2022/components/tiles/tile-handle.directive.mjs
import { Directive, inject, NgZone } from '@angular/core';
import { tuiGetActualTarget, tuiIsElement } from '@taiga-ui/cdk/utils/dom';
import { TuiTile } from './tile.component';
import * as i0 from "@angular/core";
class TuiTileHandle {
    constructor() {
        this.zone = inject(NgZone);
        this.tile = inject(TuiTile);
        this.x = NaN;
        this.y = NaN;
    }
    onPointer(x = NaN, y = NaN) {
        if (!Number.isNaN(x) || !Number.isNaN(this.x)) {
            this.zone.run(() => {
                const { left, top } = this.tile.element.getBoundingClientRect();
                this.x = x - left;
                this.y = y - top;
                this.tile.onDrag([NaN, NaN]);
            });
        }
    }
    onMove(x, y) {
        if (!Number.isNaN(this.x)) {
            this.zone.run(() => {
                this.tile.onDrag([x - this.x, y - this.y]);
            });
        }
    }
    onStart(event) {
        const target = tuiGetActualTarget(event);
        const { x, y, pointerId } = event;
        if (tuiIsElement(target)) {
            target.releasePointerCapture(pointerId);
        }
        this.onPointer(x, y);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTileHandle, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTileHandle, isStandalone: true, selector: "[tuiTileHandle]", host: { listeners: { "pointerdown.zoneless.prevent": "onStart($event)", "document:pointerup.zoneless": "onPointer()", "document:pointermove.zoneless": "onMove($event.x, $event.y)" } }, ngImport: i0 }); }
}
export { TuiTileHandle };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTileHandle, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiTileHandle]',
                    host: {
                        '(pointerdown.zoneless.prevent)': 'onStart($event)',
                        '(document:pointerup.zoneless)': 'onPointer()',
                        '(document:pointermove.zoneless)': 'onMove($event.x, $event.y)',
                    },
                }]
        }] });
