// @taiga-ui/core/esm2022/components/data-list/option/option.directive.mjs
import { createComponent, Directive, EnvironmentInjector, forwardRef, inject, INJECTOR, Input, signal, ViewContainerRef, } from '@angular/core';
import { TUI_IS_MOBILE } from '@taiga-ui/cdk/tokens';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { tuiIsNativeFocused } from '@taiga-ui/cdk/utils/focus';
import { TuiDropdownDirective } from '@taiga-ui/core/directives/dropdown';
import { TuiWithIcons } from '@taiga-ui/core/directives/icons';
import { TuiDataListComponent } from '../data-list.component';
import { TUI_DATA_LIST_HOST } from '../data-list.tokens';
import { TUI_OPTION_CONTENT } from './option-content';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/icons";
// TODO(v5): rename `TuiOptionNew` => `TuiOption` & remove [new] from selector
// TODO: Consider all use cases for aria roles
class TuiOptionNew {
    constructor() {
        this.vcr = inject(ViewContainerRef);
        this.isMobile = inject(TUI_IS_MOBILE);
        this.el = tuiInjectElement();
        this.dataList = inject(forwardRef(() => TuiDataListComponent), { optional: true });
        this.content = inject(TUI_OPTION_CONTENT, {
            optional: true,
        });
        this.ref = this.content &&
            createComponent(this.content, {
                environmentInjector: inject(EnvironmentInjector),
                elementInjector: inject(INJECTOR),
                hostElement: tuiInjectElement(),
            });
        this.dropdown = inject(TuiDropdownDirective, {
            self: true,
            optional: true,
        })?.ref;
        this.disabled = false;
        if (this.ref) {
            this.vcr.insert(this.ref.hostView);
            this.ref.changeDetectorRef.detectChanges();
        }
    }
    // Preventing focus loss upon focused option removal
    ngOnDestroy() {
        this.dataList?.handleFocusLossIfNecessary(this.el);
    }
    onMouseMove() {
        if (!this.isMobile &&
            !tuiIsNativeFocused(this.el) &&
            this.dataList &&
            this.el.closest('[tuiDataListDropdownManager]')) {
            this.el.focus({ preventScroll: true });
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiOptionNew, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiOptionNew, isStandalone: true, selector: "button[tuiOption][new], a[tuiOption][new], label[tuiOption][new]", inputs: { disabled: "disabled" }, host: { attributes: { "type": "button", "role": "option" }, listeners: { "mousemove.zoneless": "onMouseMove()" }, properties: { "attr.disabled": "disabled || null", "class._with-dropdown": "dropdown?.()" } }, hostDirectives: [{ directive: i1.TuiWithIcons }], ngImport: i0 }); }
}
export { TuiOptionNew };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiOptionNew, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'button[tuiOption][new], a[tuiOption][new], label[tuiOption][new]',
                    hostDirectives: [TuiWithIcons],
                    host: {
                        type: 'button',
                        role: 'option',
                        '[attr.disabled]': 'disabled || null',
                        '[class._with-dropdown]': 'dropdown?.()',
                        '(mousemove.zoneless)': 'onMouseMove()',
                    },
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { disabled: [{
                type: Input
            }] } });
// TODO(v5): remove [new] from selector
class TuiOptionWithValue {
    constructor() {
        this.host = inject(TUI_DATA_LIST_HOST, {
            optional: true,
        });
        this.disabled = false;
        this.value = signal(undefined);
    }
    // TODO(v5): use `input.required<T>()` to remove `undefined` from `this.value()`
    set valueSetter(x) {
        this.value.set(x);
    }
    onClick(value = this.value()) {
        if (this.host?.handleOption && value !== undefined) {
            this.host.handleOption(value);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiOptionWithValue, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiOptionWithValue, isStandalone: true, selector: "button[tuiOption][value][new], a[tuiOption][value][new], label[tuiOption][value][new]", inputs: { disabled: "disabled", valueSetter: ["value", "valueSetter"] }, host: { listeners: { "click": "onClick()" } }, ngImport: i0 }); }
}
export { TuiOptionWithValue };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiOptionWithValue, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'button[tuiOption][value][new], a[tuiOption][value][new], label[tuiOption][value][new]',
                    host: {
                        '(click)': 'onClick()',
                    },
                }]
        }], propDecorators: { disabled: [{
                type: Input
            }], valueSetter: [{
                type: Input,
                args: [{ alias: 'value', required: true }]
            }] } });
