// @taiga-ui/cdk/esm2022/directives/obscured/obscured.directive.mjs
import { Directive, inject, Input, Output } from '@angular/core';
import { TuiActiveZone } from '@taiga-ui/cdk/directives/active-zone';
import { tuiIfMap } from '@taiga-ui/cdk/observables';
import { BehaviorSubject, map } from 'rxjs';
import { TuiObscuredService } from './obscured.service';
import * as i0 from "@angular/core";
/**
 * Directive that monitors element visibility
 */
class TuiObscured {
    constructor() {
        this.activeZone = inject(TuiActiveZone, { optional: true });
        this.enabled$ = new BehaviorSubject(false);
        this.obscured$ = inject(TuiObscuredService, { self: true }).pipe(
        // TODO: Refactor so that dropdowns and dialogs work properly without hacks
        map((by) => !!by?.every((el) => el.closest('tui-dialogs') || !this.activeZone?.contains(el))));
        this.tuiObscured = this.enabled$.pipe(tuiIfMap(() => this.obscured$));
    }
    set tuiObscuredEnabled(enabled) {
        this.enabled$.next(enabled);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiObscured, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiObscured, isStandalone: true, selector: "[tuiObscured]", inputs: { tuiObscuredEnabled: "tuiObscuredEnabled" }, outputs: { tuiObscured: "tuiObscured" }, providers: [TuiObscuredService], ngImport: i0 }); }
}
export { TuiObscured };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiObscured, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiObscured]',
                    providers: [TuiObscuredService],
                }]
        }], propDecorators: { tuiObscured: [{
                type: Output
            }], tuiObscuredEnabled: [{
                type: Input
            }] } });
