// @taiga-ui/kit/esm2022/components/slider/helpers/key-steps.mjs
import { tuiRound } from '@taiga-ui/cdk/utils/math';
/**
 * Used as a limit for eliminating JS issues with floating point math
 */
export const TUI_FLOATING_PRECISION = 7;
function tuiFindKeyStepsBoundariesByFn(keySteps, fn) {
    const keyStepUpperIndex = keySteps.findIndex((keyStep, i) => i && fn(keyStep));
    const lowerStep = keySteps[keyStepUpperIndex - 1] || keySteps[0];
    const upperStep = keySteps[keyStepUpperIndex] ||
        keySteps[keySteps.length - 1] || [0, 0];
    return [lowerStep, upperStep];
}
export function tuiPercentageToKeyStepValue(valuePercentage, keySteps) {
    const [[lowerStepPercent, lowerStepValue], [upperStepPercent, upperStepValue]] = tuiFindKeyStepsBoundariesByFn(keySteps, ([keyStepPercentage, _]) => valuePercentage <= keyStepPercentage);
    const ratio = (valuePercentage - lowerStepPercent) / (upperStepPercent - lowerStepPercent);
    const controlValue = (upperStepValue - lowerStepValue) * ratio + lowerStepValue;
    return tuiRound(controlValue, TUI_FLOATING_PRECISION);
}
export function tuiKeyStepValueToPercentage(value, keySteps) {
    const [[lowerStepPercent, lowerStepValue], [upperStepPercent, upperStepValue]] = tuiFindKeyStepsBoundariesByFn(keySteps, ([_, keyStepValue]) => value <= keyStepValue);
    const ratio = (value - lowerStepValue) / (upperStepValue - lowerStepValue) || 0;
    return (upperStepPercent - lowerStepPercent) * ratio + lowerStepPercent;
}
export function tuiCreateKeyStepsTransformer(keySteps, slider) {
    return new (class {
        fromControlValue(controlValue) {
            const newValuePercentage = tuiKeyStepValueToPercentage(controlValue, keySteps);
            return (newValuePercentage * (slider.max - slider.min)) / 100 + slider.min;
        }
        toControlValue(nativeValue) {
            const valueRatio = (nativeValue - slider.min) / (slider.max - slider.min) || 0;
            return tuiPercentageToKeyStepValue(valueRatio * 100, keySteps);
        }
    })();
}
