// @taiga-ui/core/esm2022/directives/popup/popups.component.mjs
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { tuiAsPortal, TuiPortals } from '@taiga-ui/cdk/classes';
import { TuiPopupService } from './popup.service';
import * as i0 from "@angular/core";
class TuiPopups extends TuiPortals {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPopups, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiPopups, isStandalone: true, selector: "tui-popups", providers: [tuiAsPortal(TuiPopupService)], usesInheritance: true, ngImport: i0, template: '<ng-container #viewContainer />', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
export { TuiPopups };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPopups, decorators: [{
            type: Component,
            args: [{
                    standalone: true,
                    selector: 'tui-popups',
                    template: '<ng-container #viewContainer />',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [tuiAsPortal(TuiPopupService)],
                }]
        }] });
