// @taiga-ui/kit/esm2022/components/input-number/quantum.directive.mjs
import { Directive, inject } from '@angular/core';
import { TuiValueTransformer } from '@taiga-ui/cdk/classes';
import { tuiIsSafeToRound, tuiRound } from '@taiga-ui/cdk/utils/math';
import { tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
import { tuiGetFractionPartPadded } from '@taiga-ui/core/utils/format';
import { TUI_FLOATING_PRECISION } from '@taiga-ui/kit/components/slider';
import { TUI_INPUT_NUMBER_OPTIONS } from './input-number.options';
import * as i0 from "@angular/core";
class TuiQuantumValueTransformerBase extends TuiValueTransformer {
    // eslint-disable-next-line @typescript-eslint/parameter-properties
    constructor(quantum = 0) {
        super();
        this.quantum = quantum;
        this.parent = null;
    }
    fromControlValue(controlValue) {
        return this.parent?.fromControlValue(controlValue) ?? controlValue;
    }
    toControlValue(internalValue) {
        const value = this.parent?.toControlValue(internalValue) ?? internalValue;
        return value != null &&
            this.quantum > 0 &&
            tuiIsSafeToRound(value, tuiGetFractionPartPadded(this.quantum).length)
            ? tuiRound(Math.round(value / this.quantum) * this.quantum, TUI_FLOATING_PRECISION)
            : value;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiQuantumValueTransformerBase, deps: "invalid", target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiQuantumValueTransformerBase, usesInheritance: true, ngImport: i0 }); }
}
export { TuiQuantumValueTransformerBase };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiQuantumValueTransformerBase, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: undefined }]; } });
class TuiQuantumValueTransformer extends TuiQuantumValueTransformerBase {
    constructor() {
        super(0);
        this.parent = inject(TUI_INPUT_NUMBER_OPTIONS).valueTransformer;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiQuantumValueTransformer, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiQuantumValueTransformer, isStandalone: true, selector: "[tuiInputNumber][quantum]", inputs: { quantum: "quantum" }, providers: [tuiProvide(TuiValueTransformer, TuiQuantumValueTransformer)], usesInheritance: true, ngImport: i0 }); }
}
export { TuiQuantumValueTransformer };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiQuantumValueTransformer, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiInputNumber][quantum]',
                    inputs: ['quantum'],
                    providers: [tuiProvide(TuiValueTransformer, TuiQuantumValueTransformer)],
                }]
        }], ctorParameters: function () { return []; } });
class TuiWithQuantumValueTransformer {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiWithQuantumValueTransformer, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiWithQuantumValueTransformer, isStandalone: true, hostDirectives: [{ directive: TuiQuantumValueTransformer, inputs: ["quantum", "quantum"] }], ngImport: i0 }); }
}
export { TuiWithQuantumValueTransformer };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiWithQuantumValueTransformer, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    hostDirectives: [
                        {
                            directive: TuiQuantumValueTransformer,
                            inputs: ['quantum'],
                        },
                    ],
                }]
        }] });
