// @taiga-ui/core/esm2022/components/textfield/textfield-content.directive.mjs
import { Directive, inject, TemplateRef } from '@angular/core';
import { TuiTextfieldComponent } from './textfield.component';
import * as i0 from "@angular/core";
class TuiTextfieldContent {
    constructor() {
        this.ref = inject(TuiTextfieldComponent).vcr?.createEmbeddedView(inject(TemplateRef));
    }
    ngDoCheck() {
        this.ref?.detectChanges();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextfieldContent, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTextfieldContent, isStandalone: true, selector: "ng-template[tuiTextfieldContent]", ngImport: i0 }); }
}
export { TuiTextfieldContent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextfieldContent, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-template[tuiTextfieldContent]',
                }]
        }] });
