// @taiga-ui/core/esm2022/components/textfield/textfield-dropdown.directive.mjs
import { isPlatformBrowser } from '@angular/common';
import { Directive, inject, PLATFORM_ID, TemplateRef } from '@angular/core';
import { TuiDropdownDirective } from '@taiga-ui/core/directives/dropdown';
import * as i0 from "@angular/core";
// TODO: Change selector to tuiDropdown in v5 and move to TuiDropdown
class TuiTextfieldDropdownDirective {
    constructor() {
        this.directive = inject(TuiDropdownDirective);
        this.directive.tuiDropdown = inject(TemplateRef);
        if (isPlatformBrowser(inject(PLATFORM_ID)) &&
            this.directive.el.matches(':focus-within')) {
            this.directive.toggle(true);
        }
    }
    ngOnDestroy() {
        this.directive.tuiDropdown = null;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextfieldDropdownDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTextfieldDropdownDirective, isStandalone: true, selector: "ng-template[tuiTextfieldDropdown]", ngImport: i0 }); }
}
export { TuiTextfieldDropdownDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextfieldDropdownDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-template[tuiTextfieldDropdown]',
                }]
        }], ctorParameters: function () { return []; } });
/**
 * @deprecated remove in v5
 */
class TuiWithTextfieldDropdown {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiWithTextfieldDropdown, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiWithTextfieldDropdown, isStandalone: true, ngImport: i0 }); }
}
export { TuiWithTextfieldDropdown };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiWithTextfieldDropdown, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                }]
        }] });
