// @taiga-ui/layout/esm2022/components/navigation/drawer.component.mjs
import { ChangeDetectionStrategy, Component, inject, TemplateRef, ViewChild, } from '@angular/core';
import { TuiActiveZone } from '@taiga-ui/cdk/directives/active-zone';
import { TuiAnimated } from '@taiga-ui/cdk/directives/animated';
import { tuiIsHTMLElement } from '@taiga-ui/cdk/utils/dom';
import { tuiButtonOptionsProvider } from '@taiga-ui/core/components/button';
import { TuiScrollbar } from '@taiga-ui/core/components/scrollbar';
import { TUI_DROPDOWN_COMPONENT, tuiDropdown, TuiDropdownDirective, TuiDropdownOpen, TuiWithDropdownOpen, } from '@taiga-ui/core/directives/dropdown';
import { TuiIcons } from '@taiga-ui/core/directives/icons';
import { TUI_ANIMATIONS_SPEED, TUI_COMMON_ICONS, TUI_ICON_END, } from '@taiga-ui/core/tokens';
import { tuiToAnimationOptions } from '@taiga-ui/core/utils/miscellaneous';
import { TUI_LAYOUT_ICONS } from '@taiga-ui/layout/tokens';
import { PolymorpheusOutlet } from '@taiga-ui/polymorpheus';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk/directives/active-zone";
import * as i2 from "@taiga-ui/cdk/directives/animated";
import * as i3 from "@taiga-ui/core/directives/dropdown";
class TuiDrawerComponent {
    constructor() {
        this.directive = inject(TuiDropdownDirective);
        this.animation = tuiToAnimationOptions(inject(TUI_ANIMATIONS_SPEED));
        this.top = tuiIsHTMLElement(this.directive.el.offsetParent) &&
            tuiIsHTMLElement(this.directive.el.offsetParent.offsetParent)
            ? this.directive.el.offsetParent.getBoundingClientRect().bottom -
                Math.max(this.directive.el.offsetParent.offsetParent.getBoundingClientRect().top, 0)
            : 0;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDrawerComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiDrawerComponent, isStandalone: true, selector: "ng-component", host: { attributes: { "tuiTheme": "dark" }, properties: { "style.top.px": "top" } }, hostDirectives: [{ directive: i1.TuiActiveZone }, { directive: i2.TuiAnimated }], ngImport: i0, template: `
        <tui-scrollbar [style.height.%]="100">
            <ng-container *polymorpheusOutlet="directive._content()" />
        </tui-scrollbar>
    `, isInline: true, styles: [":host{position:fixed;left:0;right:0;bottom:0;background:var(--tui-theme-color, #000);inline-size:14.375rem;color:var(--tui-text-primary);pointer-events:auto;--tui-from: translateX(-100%)}:host.tui-enter,:host.tui-leave{animation-name:tuiSlide}:host ::ng-deep tui-data-list [tuiOption]{justify-content:flex-start;gap:.75rem;margin:0;font:var(--tui-font-text-ui-s);font-weight:700}:host ::ng-deep tui-data-list tui-opt-group:before{font:var(--tui-font-text-xs);text-transform:uppercase;color:var(--tui-text-tertiary)}\n"], dependencies: [{ kind: "directive", type: PolymorpheusOutlet, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { kind: "component", type: TuiScrollbar, selector: "tui-scrollbar", inputs: ["hidden"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDrawerComponent, decorators: [{
            type: Component,
            args: [{ standalone: true, imports: [PolymorpheusOutlet, TuiScrollbar], template: `
        <tui-scrollbar [style.height.%]="100">
            <ng-container *polymorpheusOutlet="directive._content()" />
        </tui-scrollbar>
    `, changeDetection: ChangeDetectionStrategy.OnPush, hostDirectives: [TuiActiveZone, TuiAnimated], host: {
                        tuiTheme: 'dark',
                        '[style.top.px]': 'top',
                    }, styles: [":host{position:fixed;left:0;right:0;bottom:0;background:var(--tui-theme-color, #000);inline-size:14.375rem;color:var(--tui-text-primary);pointer-events:auto;--tui-from: translateX(-100%)}:host.tui-enter,:host.tui-leave{animation-name:tuiSlide}:host ::ng-deep tui-data-list [tuiOption]{justify-content:flex-start;gap:.75rem;margin:0;font:var(--tui-font-text-ui-s);font-weight:700}:host ::ng-deep tui-data-list tui-opt-group:before{font:var(--tui-font-text-xs);text-transform:uppercase;color:var(--tui-text-tertiary)}\n"] }]
        }] });
class TuiDrawerDirective {
    constructor() {
        this.x = inject(TUI_COMMON_ICONS).close;
        this.icons = inject(TuiIcons);
        this.dropdown = tuiDropdown(null);
        this.open = inject(TuiDropdownOpen);
    }
    ngDoCheck() {
        this.icons.iconStart.set(this.open.tuiDropdownOpen ? this.x : '');
    }
    set template(template) {
        this.dropdown.set(template);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDrawerDirective, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiDrawerDirective, isStandalone: true, selector: "[tuiIconButton][tuiNavigationDrawer]", providers: [
            tuiButtonOptionsProvider({ appearance: 'secondary-grayscale' }),
            {
                provide: TUI_ICON_END,
                useFactory: () => inject(TUI_LAYOUT_ICONS).grid,
            },
            {
                provide: TUI_DROPDOWN_COMPONENT,
                useValue: TuiDrawerComponent,
            },
        ], viewQueries: [{ propertyName: "template", first: true, predicate: TemplateRef, descendants: true }], hostDirectives: [{ directive: i3.TuiDropdownDirective }, { directive: i3.TuiWithDropdownOpen }], ngImport: i0, template: '<ng-template><ng-content /></ng-template>', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
export { TuiDrawerDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDrawerDirective, decorators: [{
            type: Component,
            args: [{
                    standalone: true,
                    // TODO: move to host directives
                    selector: '[tuiIconButton][tuiNavigationDrawer]',
                    template: '<ng-template><ng-content /></ng-template>',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        tuiButtonOptionsProvider({ appearance: 'secondary-grayscale' }),
                        {
                            provide: TUI_ICON_END,
                            useFactory: () => inject(TUI_LAYOUT_ICONS).grid,
                        },
                        {
                            provide: TUI_DROPDOWN_COMPONENT,
                            useValue: TuiDrawerComponent,
                        },
                    ],
                    hostDirectives: [TuiDropdownDirective, TuiWithDropdownOpen],
                }]
        }], propDecorators: { template: [{
                type: ViewChild,
                args: [TemplateRef]
            }] } });
