// @taiga-ui/layout/esm2022/components/navigation/hint-aside.directive.mjs
import { computed, Directive, inject, Input, signal } from '@angular/core';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { tuiDirectiveBinding } from '@taiga-ui/cdk/utils/miscellaneous';
import { TuiDropdownDirective } from '@taiga-ui/core/directives/dropdown';
import { TuiHintDirective, tuiHintOptionsProvider } from '@taiga-ui/core/directives/hint';
import { TuiAsideComponent } from './aside.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/hint";
class TuiHintAsideDirective {
    constructor() {
        this.el = tuiInjectElement();
        this.aside = inject(TuiAsideComponent);
        this.dropdown = inject(TuiDropdownDirective, { optional: true });
        this.hint = signal(null);
        this.binding = tuiDirectiveBinding(TuiHintDirective, 'tuiHint', computed(() => {
            const hint = this.hint();
            const expanded = this.aside.expanded();
            return expanded || this.dropdown
                ? ''
                : hint || (() => this.el.textContent?.trim());
        }));
    }
    set tuiHintAside(value) {
        this.hint.set(value);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiHintAsideDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiHintAsideDirective, isStandalone: true, selector: "[tuiHintAside]", inputs: { tuiHintAside: "tuiHintAside" }, providers: [tuiHintOptionsProvider({ direction: 'right' })], hostDirectives: [{ directive: i1.TuiHintDirective }], ngImport: i0 }); }
}
export { TuiHintAsideDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiHintAsideDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiHintAside]',
                    providers: [tuiHintOptionsProvider({ direction: 'right' })],
                    hostDirectives: [TuiHintDirective],
                }]
        }], propDecorators: { tuiHintAside: [{
                type: Input
            }] } });
