// @taiga-ui/layout/esm2022/components/search/search-filter.component.mjs
import { ChangeDetectionStrategy, Component, inject, TemplateRef, ViewChild, } from '@angular/core';
import { tuiDropdown, TuiDropdownDirective, tuiDropdownOptionsProvider, TuiWithDropdownOpen, } from '@taiga-ui/core/directives/dropdown';
import { TUI_ICON_START } from '@taiga-ui/core/tokens';
import { TUI_COMMON_ICONS } from '@taiga-ui/layout/tokens';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/dropdown";
class TuiSearchFilterComponent {
    constructor() {
        this.dropdown = tuiDropdown(null);
    }
    set template(template) {
        this.dropdown.set(template);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSearchFilterComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiSearchFilterComponent, isStandalone: true, selector: "button[tuiSearchFilter]", providers: [
            tuiDropdownOptionsProvider({ align: 'right' }),
            {
                provide: TUI_ICON_START,
                useFactory: () => inject(TUI_COMMON_ICONS).filter,
            },
        ], viewQueries: [{ propertyName: "template", first: true, predicate: TemplateRef, descendants: true }], hostDirectives: [{ directive: i1.TuiDropdownDirective }, { directive: i1.TuiWithDropdownOpen }], ngImport: i0, template: '<ng-template><ng-content /></ng-template>filters', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
export { TuiSearchFilterComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSearchFilterComponent, decorators: [{
            type: Component,
            args: [{
                    standalone: true,
                    selector: 'button[tuiSearchFilter]',
                    template: '<ng-template><ng-content /></ng-template>filters',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        tuiDropdownOptionsProvider({ align: 'right' }),
                        {
                            provide: TUI_ICON_START,
                            useFactory: () => inject(TUI_COMMON_ICONS).filter,
                        },
                    ],
                    hostDirectives: [TuiDropdownDirective, TuiWithDropdownOpen],
                }]
        }], propDecorators: { template: [{
                type: ViewChild,
                args: [TemplateRef]
            }] } });
