// @taiga-ui/kit/esm2022/directives/button-select/button-select.directive.mjs
import { Directive, inject } from '@angular/core';
import { TuiControl } from '@taiga-ui/cdk/classes';
import { TuiNativeValidator } from '@taiga-ui/cdk/directives/native-validator';
import { tuiArrayToggle } from '@taiga-ui/cdk/utils/miscellaneous';
import { tuiAsDataListHost, tuiAsOptionContent, } from '@taiga-ui/core/components/data-list';
import { TuiWithTextfieldDropdown } from '@taiga-ui/core/components/textfield';
import { TuiDropdownDirective, tuiDropdownOpen, tuiDropdownOptionsProvider, TuiWithDropdownOpen, } from '@taiga-ui/core/directives/dropdown';
import { TUI_ITEMS_HANDLERS } from '@taiga-ui/core/directives/items-handlers';
import { TuiSelectOption } from '@taiga-ui/kit/components/select';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/dropdown";
import * as i2 from "@taiga-ui/cdk/directives/native-validator";
import * as i3 from "@taiga-ui/core/components/textfield";
class TuiButtonSelect extends TuiControl {
    constructor() {
        super(...arguments);
        this.open = tuiDropdownOpen();
        this.handlers = inject(TUI_ITEMS_HANDLERS);
        this.size = 's';
    }
    handleOption(option) {
        if (Array.isArray(this.control.value)) {
            const result = tuiArrayToggle(this.control.value, option, this.handlers.identityMatcher());
            this.onChange(result);
        }
        else {
            this.onChange(option);
            this.open.set(false);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiButtonSelect, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiButtonSelect, isStandalone: true, selector: "button[tuiButtonSelect]", providers: [
            tuiAsOptionContent(TuiSelectOption),
            tuiAsDataListHost(TuiButtonSelect),
            tuiDropdownOptionsProvider({ align: 'right' }),
        ], usesInheritance: true, hostDirectives: [{ directive: i1.TuiDropdownDirective }, { directive: i2.TuiNativeValidator }, { directive: i1.TuiWithDropdownOpen }, { directive: i3.TuiWithTextfieldDropdown }], ngImport: i0 }); }
}
export { TuiButtonSelect };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiButtonSelect, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'button[tuiButtonSelect]',
                    providers: [
                        tuiAsOptionContent(TuiSelectOption),
                        tuiAsDataListHost(TuiButtonSelect),
                        tuiDropdownOptionsProvider({ align: 'right' }),
                    ],
                    hostDirectives: [
                        TuiDropdownDirective,
                        TuiNativeValidator,
                        TuiWithDropdownOpen,
                        TuiWithTextfieldDropdown,
                    ],
                }]
        }] });
