// @taiga-ui/cdk/esm2022/directives/font-size/font-size.directive.mjs
import { DOCUMENT, isPlatformBrowser } from '@angular/common';
import { DestroyRef, Directive, inject, InjectionToken, PLATFORM_ID } from '@angular/core';
import { EMPTY_FUNCTION } from '@taiga-ui/cdk/constants';
import { tuiFontSizeWatcher } from '@taiga-ui/font-watcher';
import * as i0 from "@angular/core";
export const TUI_FONT_SIZE_HANDLER = new InjectionToken(ngDevMode ? 'TUI_FONT_SIZE_HANDLER' : '');
class TuiFontSize {
    constructor() {
        this.handler = inject(TUI_FONT_SIZE_HANDLER, { optional: true });
        this.enabled = !inject(TuiFontSize, { optional: true, skipSelf: true }) &&
            isPlatformBrowser(inject(PLATFORM_ID)) &&
            typeof ResizeObserver !== 'undefined';
        this.nothing = inject(DestroyRef).onDestroy(this.enabled && this.handler
            ? tuiFontSizeWatcher(this.handler, inject(DOCUMENT).createElement('iframe'))
            : EMPTY_FUNCTION);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiFontSize, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiFontSize, isStandalone: true, ngImport: i0 }); }
}
export { TuiFontSize };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiFontSize, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                }]
        }] });
