// @taiga-ui/cdk/esm2022/directives/auto-focus/autofocus.options.mjs
import { ElementRef, InjectionToken, NgZone, Renderer2 } from '@angular/core';
import { WA_ANIMATION_FRAME, WA_WINDOW } from '@ng-web-apis/common';
import { TUI_IS_IOS } from '@taiga-ui/cdk/tokens';
import { tuiCreateOptions } from '@taiga-ui/cdk/utils';
import { TuiDefaultAutofocusHandler } from './handlers/default.handler';
import { TuiIosAutofocusHandler } from './handlers/ios.handler';
export const [TUI_AUTOFOCUS_OPTIONS, tuiAutoFocusOptionsProvider] = tuiCreateOptions({
    delay: NaN,
    query: 'input, textarea, select, [contenteditable]',
    preventScroll: false,
});
export const TUI_AUTOFOCUS_HANDLER = new InjectionToken(ngDevMode ? 'TUI_AUTOFOCUS_HANDLER' : '');
export const TUI_AUTOFOCUS_PROVIDERS = [
    {
        provide: TUI_AUTOFOCUS_HANDLER,
        deps: [
            ElementRef,
            WA_ANIMATION_FRAME,
            Renderer2,
            NgZone,
            WA_WINDOW,
            TUI_IS_IOS,
            TUI_AUTOFOCUS_OPTIONS,
        ],
        // eslint-disable-next-line @typescript-eslint/max-params,max-params
        useFactory: (el, animationFrame$, renderer, zone, win, isIos, options) => isIos
            ? new TuiIosAutofocusHandler(el, renderer, zone, win, options)
            : new TuiDefaultAutofocusHandler(el, animationFrame$, zone, options),
    },
];
