// @taiga-ui/cdk/fesm2022/taiga-ui-cdk-directives-transitioned.mjs
import * as i0 from '@angular/core';
import { inject, NgZone, Directive } from '@angular/core';
import { tuiInjectElement } from '@taiga-ui/cdk/utils';

class TuiTransitioned {
    constructor() {
        const el = tuiInjectElement();
        inject(NgZone).runOutsideAngular(() => {
            setTimeout(() => {
                el.style.transition = '';
            });
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTransitioned, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTransitioned, isStandalone: true, selector: "[tuiTransitioned]", host: { properties: { "style.transition": "\"none\"" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTransitioned, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiTransitioned]',
                    host: {
                        '[style.transition]': '"none"',
                    },
                }]
        }], ctorParameters: function () { return []; } });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiTransitioned };
