// @taiga-ui/kit/esm2022/components/range/range-change.directive.mjs
import { DOCUMENT } from '@angular/common';
import { Directive, EventEmitter, inject, Output } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { tuiTypedFromEvent } from '@taiga-ui/cdk/observables';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { tuiClamp, tuiRound } from '@taiga-ui/cdk/utils/math';
import { TUI_FLOATING_PRECISION } from '@taiga-ui/kit/components/slider';
import { map, repeat, startWith, switchMap, takeUntil, tap } from 'rxjs';
import { TuiRange } from './range.component';
import * as i0 from "@angular/core";
class TuiRangeChange {
    constructor() {
        this.doc = inject(DOCUMENT);
        this.el = tuiInjectElement();
        this.range = inject(TuiRange);
        // TODO(v5): use 'start' | 'end' instead
        this.activeThumbChange = new EventEmitter();
        let activeThumb;
        tuiTypedFromEvent(this.el, 'pointerdown', {
            passive: true,
            capture: true,
        })
            .pipe(tap(({ clientX, target, pointerId }) => {
            activeThumb = this.detectActiveThumb(clientX, target);
            this.range.slidersRefs
                .get(activeThumb === 'start' ? 0 : 1)
                ?.nativeElement.setPointerCapture(pointerId);
            // TODO(v5): remove backward compatibility
            this.activeThumbChange.emit(activeThumb === 'start' ? 'left' : 'right');
            if (this.range.focusable) {
                this.el.focus();
            }
        }), switchMap((event) => tuiTypedFromEvent(this.doc, 'pointermove').pipe(startWith(event))), map(({ clientX }) => this.getFractionFromEvents(clientX ?? 0)), takeUntil(tuiTypedFromEvent(this.doc, 'pointerup', { passive: true })), repeat(), takeUntilDestroyed())
            .subscribe((fraction) => {
            const value = this.range.toValue(fraction);
            this.range.processValue(value, activeThumb === 'end');
        });
    }
    getFractionFromEvents(clickClientX) {
        const { left, right, width } = this.el.getBoundingClientRect();
        const start = this.el.matches('[dir="rtl"] :scope') ? right : left;
        const value = Math.abs(tuiClamp(clickClientX, left, right) - start);
        return tuiRound(value / width, TUI_FLOATING_PRECISION);
    }
    detectActiveThumb(clientX, target) {
        const [startSliderRef, endSliderRef] = this.range.slidersRefs;
        switch (target) {
            case endSliderRef?.nativeElement:
                return 'end';
            case startSliderRef?.nativeElement:
                return 'start';
            default:
                return this.findNearestActiveThumb(clientX);
        }
    }
    findNearestActiveThumb(clientX) {
        const fraction = this.getFractionFromEvents(clientX);
        const deltaStart = fraction * 100 - this.range.start();
        const deltaEnd = fraction * 100 - 100 + this.range.end();
        return Math.abs(deltaStart) > Math.abs(deltaEnd) ||
            deltaEnd > 0 ||
            (this.range.start() === 0 && this.range.end() === 100)
            ? 'end'
            : 'start';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiRangeChange, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiRangeChange, isStandalone: true, outputs: { activeThumbChange: "activeThumbChange" }, ngImport: i0 }); }
}
export { TuiRangeChange };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiRangeChange, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { activeThumbChange: [{
                type: Output
            }] } });
