// @taiga-ui/layout/esm2022/components/app-bar/app-bar-size.directive.mjs
import { Directive, inject } from '@angular/core';
import { toSignal } from '@angular/core/rxjs-interop';
import { tuiDirectiveBinding } from '@taiga-ui/cdk/utils/miscellaneous';
import { TuiBreakpointService } from '@taiga-ui/core/services';
import { map } from 'rxjs';
import { TuiAppBarComponent } from './app-bar.component';
import * as i0 from "@angular/core";
// TODO: Make size automatic based on tuiPlatform in v5
class TuiAppBarSizeDirective {
    constructor() {
        this.size = tuiDirectiveBinding(TuiAppBarComponent, 'size', toSignal(inject(TuiBreakpointService).pipe(map((breakpoint) => (breakpoint === 'mobile' ? 'm' : 'l'))), { initialValue: 'm' }));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAppBarSizeDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiAppBarSizeDirective, isStandalone: true, selector: "tui-app-bar[tuiAppBarSize]", ngImport: i0 }); }
}
export { TuiAppBarSizeDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAppBarSizeDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'tui-app-bar[tuiAppBarSize]',
                }]
        }] });
