// @taiga-ui/kit/esm2022/components/accordion/accordion.directive.mjs
import { ContentChildren, DestroyRef, Directive, inject, Input, } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { EMPTY_QUERY } from '@taiga-ui/cdk/constants';
import { tuiQueryListChanges } from '@taiga-ui/cdk/observables';
import { tuiIsPresent } from '@taiga-ui/cdk/utils/miscellaneous';
import { TuiGroup } from '@taiga-ui/core/directives/group';
import { filter, identity, map, merge, pairwise, switchMap } from 'rxjs';
import { TuiAccordionItem } from './accordion-item.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/group";
class TuiAccordionDirective {
    constructor() {
        this.destroyRef = inject(DestroyRef);
        this.accordionItems = EMPTY_QUERY;
        this.closeOthers = true;
        // Not using DI options to avoid changed defaults spilling to content
        const group = inject(TuiGroup);
        group.orientation = 'vertical';
        group.collapsed = true;
    }
    ngAfterContentInit() {
        const { accordionItems } = this;
        const rows$ = tuiQueryListChanges(accordionItems);
        const newOpenRow$ = rows$.pipe(pairwise(), map(([previous, current]) => current.find((item) => !previous.includes(item) && item.open)), filter(tuiIsPresent));
        const rowsOpen$ = merge(rows$.pipe(switchMap((rows) => merge(...rows.map((row) => row.openChange.pipe(filter(identity), map(() => row)))))), newOpenRow$).pipe(filter(() => this.closeOthers), takeUntilDestroyed(this.destroyRef));
        rowsOpen$.subscribe((currentRow) => {
            accordionItems.forEach((row) => {
                if (currentRow !== row) {
                    row.close();
                }
            });
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAccordionDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiAccordionDirective, isStandalone: true, selector: "tui-accordion", inputs: { closeOthers: "closeOthers" }, queries: [{ propertyName: "accordionItems", predicate: TuiAccordionItem }], hostDirectives: [{ directive: i1.TuiGroup, inputs: ["rounded", "rounded"] }], ngImport: i0 }); }
}
export { TuiAccordionDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAccordionDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'tui-accordion',
                    hostDirectives: [
                        {
                            directive: TuiGroup,
                            inputs: ['rounded'],
                        },
                    ],
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { accordionItems: [{
                type: ContentChildren,
                args: [TuiAccordionItem]
            }], closeOthers: [{
                type: Input
            }] } });
