// @taiga-ui/kit/esm2022/components/input-number/step/input-number-step.service.mjs
import { DOCUMENT } from '@angular/common';
import { DestroyRef, inject, Injectable } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { expand, fromEvent, map, merge, share, Subject, switchMap, takeUntil, timer, } from 'rxjs';
import * as i0 from "@angular/core";
const INITIAL_DELAY = 300;
const DELAY_DECREMENT = 15;
const MIN_DELAY = 100;
function getDelay(index) {
    return Math.max(INITIAL_DELAY - index * DELAY_DECREMENT, MIN_DELAY);
}
class TuiInputNumberStepService {
    constructor() {
        this.doc = inject(DOCUMENT);
        this.destroyRef = inject(DestroyRef);
        this.start$ = new Subject();
        this.stop$ = merge(fromEvent(this.doc, 'pointerup'), fromEvent(this.doc, 'pointerleave'), fromEvent(this.doc, 'pointercancel'));
        this.steps$ = this.start$.pipe(switchMap((value) => timer(INITIAL_DELAY).pipe(expand((_, index) => timer(getDelay(index))), map(() => value), takeUntil(this.stop$))), takeUntilDestroyed(this.destroyRef), share());
    }
    next(value) {
        this.start$.next(value);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputNumberStepService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputNumberStepService }); }
}
export { TuiInputNumberStepService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputNumberStepService, decorators: [{
            type: Injectable
        }] });
