// @taiga-ui/kit/esm2022/components/textarea/textarea-limit.directive.mjs
import { ChangeDetectionStrategy, Component, Directive, inject, Input, signal, ViewContainerRef, } from '@angular/core';
import { NG_VALIDATORS, Validators, } from '@angular/forms';
import { tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
import { TUI_TEXTFIELD_OPTIONS, TuiTextfieldComponent, } from '@taiga-ui/core/components/textfield';
import { injectContext, PolymorpheusComponent } from '@taiga-ui/polymorpheus';
import { tuiTextareaOptionsProvider } from './textarea.options';
import * as i0 from "@angular/core";
class TuiTextareaLimitComponent {
    constructor() {
        this.limit = inject(TuiTextareaLimit).limit;
        this.context = injectContext();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextareaLimitComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiTextareaLimitComponent, isStandalone: true, selector: "ng-component", ngImport: i0, template: `
        <span [textContent]="context.$implicit.slice(0, limit())"></span>
        <span
            style="background: linear-gradient(transparent 0.25rem, var(--tui-status-negative-pale) 0.25rem, var(--tui-status-negative-pale) 100%)"
            [textContent]="context.$implicit.slice(limit())"
        ></span>
    `, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
export { TuiTextareaLimitComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextareaLimitComponent, decorators: [{
            type: Component,
            args: [{
                    standalone: true,
                    template: `
        <span [textContent]="context.$implicit.slice(0, limit())"></span>
        <span
            style="background: linear-gradient(transparent 0.25rem, var(--tui-status-negative-pale) 0.25rem, var(--tui-status-negative-pale) 100%)"
            [textContent]="context.$implicit.slice(limit())"
        ></span>
    `,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });
class TuiTextareaCounterComponent {
    constructor() {
        this.limit = signal(0);
        this.length = signal(0);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextareaCounterComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiTextareaCounterComponent, isStandalone: true, selector: "ng-component", ngImport: i0, template: '{{ length() }} / {{ limit() }}', isInline: true, styles: [":host{z-index:1;inline-size:100%;order:2;text-align:end;pointer-events:none;padding-block-end:.75rem;font:var(--tui-font-text-ui-xs);color:var(--tui-text-secondary)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
export { TuiTextareaCounterComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextareaCounterComponent, decorators: [{
            type: Component,
            args: [{ standalone: true, template: '{{ length() }} / {{ limit() }}', changeDetection: ChangeDetectionStrategy.OnPush, styles: [":host{z-index:1;inline-size:100%;order:2;text-align:end;pointer-events:none;padding-block-end:.75rem;font:var(--tui-font-text-ui-xs);color:var(--tui-text-secondary)}\n"] }]
        }] });
const COMPONENT = new PolymorpheusComponent(TuiTextareaLimitComponent);
class TuiTextareaLimit {
    constructor() {
        this.textfield = inject(TuiTextfieldComponent);
        this.ref = inject(ViewContainerRef).createComponent(TuiTextareaCounterComponent);
        this.size = inject(TUI_TEXTFIELD_OPTIONS).size;
        this.limit = signal(0);
    }
    // TODO: Use signal inputs in v5
    set limitSetter(limit) {
        this.limit.set(limit);
    }
    ngDoCheck() {
        this.ref.instance.length.set(this.textfield.value().length);
        this.ref.instance.limit.set(this.limit());
    }
    validate(control) {
        return Validators.maxLength(this.limit())(control);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextareaLimit, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTextareaLimit, isStandalone: true, selector: "[tuiTextarea][limit]", inputs: { limitSetter: ["limit", "limitSetter"] }, host: { properties: { "style.border-block-end-width.rem": "size() === \"l\" ? 1.875 : 1.75" } }, providers: [
            tuiProvide(NG_VALIDATORS, TuiTextareaLimit, true),
            tuiTextareaOptionsProvider({ content: COMPONENT }),
        ], ngImport: i0 }); }
}
export { TuiTextareaLimit };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextareaLimit, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiTextarea][limit]',
                    providers: [
                        tuiProvide(NG_VALIDATORS, TuiTextareaLimit, true),
                        tuiTextareaOptionsProvider({ content: COMPONENT }),
                    ],
                    host: {
                        '[style.border-block-end-width.rem]': 'size() === "l" ? 1.875 : 1.75',
                    },
                }]
        }], propDecorators: { limitSetter: [{
                type: Input,
                args: ['limit']
            }] } });
