// @taiga-ui/kit/esm2022/components/shrink-wrap/shrink-wrap.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
class TuiShrinkWrapDirective {
    constructor() {
        this.tuiShrinkWrap = '';
    }
    get size() {
        return `calc(${this.tuiShrinkWrap || '100%'} + var(--t-shrink-wrap, 0px))`;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiShrinkWrapDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiShrinkWrapDirective, isStandalone: true, selector: "[tuiShrinkWrap]:not([tuiToast])", inputs: { tuiShrinkWrap: "tuiShrinkWrap" }, host: { attributes: { "tuiShrinkWrap": "" }, properties: { "style.max-inline-size": "size" } }, ngImport: i0 }); }
}
export { TuiShrinkWrapDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiShrinkWrapDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiShrinkWrap]:not([tuiToast])',
                    host: {
                        tuiShrinkWrap: '',
                        '[style.max-inline-size]': 'size',
                    },
                }]
        }], propDecorators: { tuiShrinkWrap: [{
                type: Input
            }] } });
