// @taiga-ui/core/esm2022/directives/appearance/appearance.directive.mjs
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, computed, Directive, effect, inject, Input, signal, ViewEncapsulation, } from '@angular/core';
import { TUI_ALLOW_SIGNAL_WRITES } from '@taiga-ui/cdk/constants';
import { TuiTransitioned } from '@taiga-ui/cdk/directives/transitioned';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { tuiIsString, tuiWithStyles } from '@taiga-ui/cdk/utils/miscellaneous';
import { TUI_APPEARANCE_OPTIONS } from './appearance.options';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk/directives/transitioned";
class TuiAppearanceStyles {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAppearanceStyles, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiAppearanceStyles, isStandalone: true, selector: "ng-component", host: { classAttribute: "tui-appearance" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiAppearance]{transition-property:all;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;position:relative;-webkit-appearance:none;appearance:none;outline:.125rem solid transparent;outline-offset:-.125rem;transition-property:color,background-color,opacity,box-shadow,border-color,border-radius,filter}[tuiAppearance]:before,[tuiAppearance]:after{transition-duration:inherit;transition-timing-function:ease-in-out}[tuiAppearance]:focus-visible:not([data-focus=false]){outline-color:var(--tui-border-focus)}[tuiAppearance][data-focus=true]{outline-color:var(--tui-border-focus)}[tuiAppearance][tuiWrapper]:not(._focused):has(:focus-visible),[tuiAppearance][tuiWrapper]._focused{outline-color:var(--tui-border-focus)}[tuiAppearance]:disabled:not([data-state]),[tuiAppearance][data-state=disabled]{cursor:initial;opacity:var(--tui-disabled-opacity)}[tuiAppearance][tuiWrapper]:disabled:not([data-state]),[tuiAppearance][tuiWrapper][data-state=disabled]{cursor:initial;opacity:var(--tui-disabled-opacity)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAppearanceStyles, decorators: [{
            type: Component,
            args: [{ standalone: true, template: '', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        class: 'tui-appearance',
                    }, styles: ["[tuiAppearance]{transition-property:all;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;position:relative;-webkit-appearance:none;appearance:none;outline:.125rem solid transparent;outline-offset:-.125rem;transition-property:color,background-color,opacity,box-shadow,border-color,border-radius,filter}[tuiAppearance]:before,[tuiAppearance]:after{transition-duration:inherit;transition-timing-function:ease-in-out}[tuiAppearance]:focus-visible:not([data-focus=false]){outline-color:var(--tui-border-focus)}[tuiAppearance][data-focus=true]{outline-color:var(--tui-border-focus)}[tuiAppearance][tuiWrapper]:not(._focused):has(:focus-visible),[tuiAppearance][tuiWrapper]._focused{outline-color:var(--tui-border-focus)}[tuiAppearance]:disabled:not([data-state]),[tuiAppearance][data-state=disabled]{cursor:initial;opacity:var(--tui-disabled-opacity)}[tuiAppearance][tuiWrapper]:disabled:not([data-state]),[tuiAppearance][tuiWrapper][data-state=disabled]{cursor:initial;opacity:var(--tui-disabled-opacity)}\n"] }]
        }] });
class TuiAppearance {
    constructor() {
        this.cdr = inject(ChangeDetectorRef, { skipSelf: true });
        this.el = tuiInjectElement();
        this.nothing = tuiWithStyles(TuiAppearanceStyles);
        this.modes = computed((mode = this.mode()) => !mode || tuiIsString(mode) ? mode : mode.join(' '));
        // TODO: refactor to signal inputs after Angular update
        this.appearance = signal(inject(TUI_APPEARANCE_OPTIONS).appearance);
        this.state = signal(null);
        this.focus = signal(null);
        this.mode = signal(null);
        // TODO: Remove when Angular is updated
        this.update = effect(() => {
            this.mode();
            this.state();
            this.focus();
            if (this.el.matches('tui-textfield[multi]')) {
                this.cdr.detectChanges();
            }
        }, TUI_ALLOW_SIGNAL_WRITES);
    }
    set tuiAppearance(appearance) {
        this.appearance.set(appearance);
    }
    set tuiAppearanceState(state) {
        this.state.set(state);
    }
    set tuiAppearanceFocus(focus) {
        this.focus.set(focus);
    }
    set tuiAppearanceMode(mode) {
        this.mode.set(mode);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAppearance, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiAppearance, isStandalone: true, selector: "[tuiAppearance]", inputs: { tuiAppearance: "tuiAppearance", tuiAppearanceState: "tuiAppearanceState", tuiAppearanceFocus: "tuiAppearanceFocus", tuiAppearanceMode: "tuiAppearanceMode" }, host: { attributes: { "tuiAppearance": "" }, properties: { "attr.data-appearance": "appearance()", "attr.data-state": "state()", "attr.data-focus": "focus()", "attr.data-mode": "modes()" } }, hostDirectives: [{ directive: i1.TuiTransitioned }], ngImport: i0 }); }
}
export { TuiAppearance };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAppearance, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiAppearance]',
                    hostDirectives: [TuiTransitioned],
                    host: {
                        tuiAppearance: '',
                        '[attr.data-appearance]': 'appearance()',
                        '[attr.data-state]': 'state()',
                        '[attr.data-focus]': 'focus()',
                        '[attr.data-mode]': 'modes()',
                    },
                }]
        }], propDecorators: { tuiAppearance: [{
                type: Input
            }], tuiAppearanceState: [{
                type: Input
            }], tuiAppearanceFocus: [{
                type: Input
            }], tuiAppearanceMode: [{
                type: Input
            }] } });
