// @taiga-ui/core/esm2022/classes/driver.mjs
import { DestroyRef, Directive, inject, } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
import { distinctUntilChanged, merge, Observable } from 'rxjs';
import { TuiVehicle } from './vehicle';
import * as i0 from "@angular/core";
export class TuiDriver extends Observable {
}
export function tuiAsDriver(driver) {
    return tuiProvide(TuiDriver, driver, true);
}
class TuiDriverDirective {
    constructor() {
        this.destroyRef = inject(DestroyRef);
        this.drivers = inject(TuiDriver, { self: true, optional: true }) || [];
        this.vehicles = inject(TuiVehicle, {
            self: true,
            optional: true,
        });
    }
    ngAfterViewInit() {
        const vehicle = this.vehicles?.find(({ type }) => type === this.type);
        merge(...this.drivers.filter(({ type }) => type === this.type))
            .pipe(distinctUntilChanged(), takeUntilDestroyed(this.destroyRef))
            .subscribe((value) => {
            vehicle?.toggle(value);
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDriverDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiDriverDirective, ngImport: i0 }); }
}
export { TuiDriverDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDriverDirective, decorators: [{
            type: Directive
        }] });
