// @taiga-ui/core/esm2022/components/alert/alerts.component.mjs
import { AsyncPipe, NgComponentOutlet, NgFor } from '@angular/common';
import { ChangeDetectionStrategy, Component, inject, INJECTOR, Injector, ViewEncapsulation, } from '@angular/core';
import { TuiAnimatedParent } from '@taiga-ui/cdk/directives/animated';
import { TuiMapperPipe } from '@taiga-ui/cdk/pipes/mapper';
// eslint-disable-next-line no-restricted-imports
import { POLYMORPHEUS_CONTEXT } from '@taiga-ui/polymorpheus';
import { combineLatest, identity, map, of } from 'rxjs';
import { TUI_ALERTS, TUI_ALERTS_GROUPED } from './alert.tokens';
import * as i0 from "@angular/core";
class TuiAlerts {
    constructor() {
        this.injector = inject(INJECTOR);
        this.alerts$ = inject(TUI_ALERTS_GROUPED);
        this.trackBy = identity;
        this.mapper = (useValue) => Injector.create({
            providers: [
                {
                    provide: POLYMORPHEUS_CONTEXT,
                    useValue,
                },
            ],
            parent: this.injector,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAlerts, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiAlerts, isStandalone: true, selector: "tui-alerts", providers: [
            {
                provide: TUI_ALERTS_GROUPED,
                useFactory: () => combineLatest([
                    of(new Map()),
                    inject(TUI_ALERTS),
                ]).pipe(map(([map, alerts]) => {
                    map.forEach((_, key) => map.set(key, []));
                    alerts.forEach((alert) => {
                        const key = alert.component.component;
                        const value = map.get(key) || [];
                        map.set(key, [...value, alert]);
                    });
                    return Array.from(map.values());
                })),
            },
        ], ngImport: i0, template: "<div\n    *ngFor=\"let group of alerts$ | async; trackBy: trackBy\"\n    tuiAnimatedParent\n    class=\"t-wrapper\"\n>\n    <ng-container\n        *ngFor=\"let item of group\"\n        [ngComponentOutlet]=\"item.component.component\"\n        [ngComponentOutletInjector]=\"item | tuiMapper: mapper\"\n    />\n</div>\n", styles: ["tui-alerts>.t-wrapper{position:fixed;top:0;left:0;inline-size:100%;display:flex;flex-direction:column;pointer-events:none;box-sizing:border-box;block-size:100%;padding:env(safe-area-inset-top) 0 env(safe-area-inset-bottom)}tui-alerts>.t-wrapper>*{pointer-events:auto}\n"], dependencies: [{ kind: "pipe", type: AsyncPipe, name: "async" }, { kind: "directive", type: NgComponentOutlet, selector: "[ngComponentOutlet]", inputs: ["ngComponentOutlet", "ngComponentOutletInputs", "ngComponentOutletInjector", "ngComponentOutletContent", "ngComponentOutletNgModule", "ngComponentOutletNgModuleFactory"] }, { kind: "directive", type: NgFor, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: TuiAnimatedParent, selector: "[tuiAnimatedParent]" }, { kind: "pipe", type: TuiMapperPipe, name: "tuiMapper" }], changeDetection: i0.ChangeDetectionStrategy.Default, encapsulation: i0.ViewEncapsulation.None }); }
}
export { TuiAlerts };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAlerts, decorators: [{
            type: Component,
            args: [{ standalone: true, selector: 'tui-alerts', imports: [AsyncPipe, NgComponentOutlet, NgFor, TuiAnimatedParent, TuiMapperPipe], encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.Default, providers: [
                        {
                            provide: TUI_ALERTS_GROUPED,
                            useFactory: () => combineLatest([
                                of(new Map()),
                                inject(TUI_ALERTS),
                            ]).pipe(map(([map, alerts]) => {
                                map.forEach((_, key) => map.set(key, []));
                                alerts.forEach((alert) => {
                                    const key = alert.component.component;
                                    const value = map.get(key) || [];
                                    map.set(key, [...value, alert]);
                                });
                                return Array.from(map.values());
                            })),
                        },
                    ], template: "<div\n    *ngFor=\"let group of alerts$ | async; trackBy: trackBy\"\n    tuiAnimatedParent\n    class=\"t-wrapper\"\n>\n    <ng-container\n        *ngFor=\"let item of group\"\n        [ngComponentOutlet]=\"item.component.component\"\n        [ngComponentOutletInjector]=\"item | tuiMapper: mapper\"\n    />\n</div>\n", styles: ["tui-alerts>.t-wrapper{position:fixed;top:0;left:0;inline-size:100%;display:flex;flex-direction:column;pointer-events:none;box-sizing:border-box;block-size:100%;padding:env(safe-area-inset-top) 0 env(safe-area-inset-bottom)}tui-alerts>.t-wrapper>*{pointer-events:auto}\n"] }]
        }] });
