// @taiga-ui/layout/esm2022/components/card/row.directive.mjs
import { ChangeDetectionStrategy, Component, Directive, ViewEncapsulation, } from '@angular/core';
import { tuiWithStyles } from '@taiga-ui/cdk/utils/miscellaneous';
import * as i0 from "@angular/core";
class TuiCardRowStyles {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiCardRowStyles, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiCardRowStyles, isStandalone: true, selector: "ng-component", host: { classAttribute: "tui-card-row" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiCardRow]{display:flex;align-items:center;gap:3rem}[tuiCardRow] tui-icon{font-size:1rem}[tuiCardRow] [tuiTitle]{font:var(--tui-font-text-s)}[tuiCardRow] [tuiTitle]>:not([tuiSubtitle]){display:flex;align-items:center;gap:.25rem}[tuiCardRow] [tuiSubtitle]{color:var(--tui-text-secondary);font:var(--tui-font-text-xs)}[tuiCardRow]>[tuiButton],[tuiCardRow]>[tuiIconButton]{margin-inline-start:auto}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiCardRowStyles, decorators: [{
            type: Component,
            args: [{ standalone: true, template: '', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        class: 'tui-card-row',
                    }, styles: ["[tuiCardRow]{display:flex;align-items:center;gap:3rem}[tuiCardRow] tui-icon{font-size:1rem}[tuiCardRow] [tuiTitle]{font:var(--tui-font-text-s)}[tuiCardRow] [tuiTitle]>:not([tuiSubtitle]){display:flex;align-items:center;gap:.25rem}[tuiCardRow] [tuiSubtitle]{color:var(--tui-text-secondary);font:var(--tui-font-text-xs)}[tuiCardRow]>[tuiButton],[tuiCardRow]>[tuiIconButton]{margin-inline-start:auto}\n"] }]
        }] });
class TuiCardRow {
    constructor() {
        this.nothing = tuiWithStyles(TuiCardRowStyles);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiCardRow, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiCardRow, isStandalone: true, selector: "[tuiCardRow]", ngImport: i0 }); }
}
export { TuiCardRow };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiCardRow, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiCardRow]',
                }]
        }] });
