// @taiga-ui/kit/esm2022/components/input-date/input-date.directive.mjs
import { computed, Directive, effect, inject, Input, signal, untracked, } from '@angular/core';
import { toSignal } from '@angular/core/rxjs-interop';
import { MaskitoDirective } from '@maskito/angular';
import { maskitoDateOptionsGenerator } from '@maskito/kit';
import { tuiAsControl, TuiControl, tuiValueTransformerFrom } from '@taiga-ui/cdk/classes';
import { TUI_ALLOW_SIGNAL_WRITES } from '@taiga-ui/cdk/constants';
import { DATE_FILLER_LENGTH, TuiDay, } from '@taiga-ui/cdk/date-time';
import { TUI_IS_MOBILE } from '@taiga-ui/cdk/tokens';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { TuiCalendar } from '@taiga-ui/core/components/calendar';
import { tuiAsOptionContent } from '@taiga-ui/core/components/data-list';
import { tuiInjectAuxiliary, TuiTextfieldDirective, tuiTextfieldIconBinding, TuiWithNativePicker, TuiWithTextfield, } from '@taiga-ui/core/components/textfield';
import { TuiDropdownAuto, tuiDropdownEnabled, tuiDropdownOpen, } from '@taiga-ui/core/directives/dropdown';
import { TUI_ITEMS_HANDLERS, TuiItemsHandlersDirective, TuiItemsHandlersValidator, } from '@taiga-ui/core/directives/items-handlers';
import { TUI_DATE_FORMAT, TUI_DEFAULT_DATE_FORMAT } from '@taiga-ui/core/tokens';
import { TuiCalendarRange } from '@taiga-ui/kit/components/calendar-range';
import { TuiSelectOption } from '@taiga-ui/kit/components/select';
import { tuiMaskito } from '@taiga-ui/kit/utils';
import { tuiWithDateFiller } from './date-filler';
import { TUI_INPUT_DATE_OPTIONS_NEW, } from './input-date.options';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/components/textfield";
import * as i2 from "@taiga-ui/core/directives/dropdown";
import * as i3 from "@taiga-ui/core/directives/items-handlers";
import * as i4 from "@maskito/angular";
export const TUI_DATE_ADAPTER = {
    DMY: 'dd/mm/yyyy',
    MDY: 'mm/dd/yyyy',
    YMD: 'yyyy/mm/dd',
};
class TuiInputDateBase extends TuiControl {
    constructor() {
        super(...arguments);
        this.calendar = tuiInjectAuxiliary((x) => x instanceof TuiCalendar || x instanceof TuiCalendarRange);
        this.el = tuiInjectElement();
        this.textfield = inject(TuiTextfieldDirective);
        this.filler = tuiWithDateFiller();
        this.mobile = inject(TUI_IS_MOBILE);
        this.open = tuiDropdownOpen();
        this.icon = tuiTextfieldIconBinding(TUI_INPUT_DATE_OPTIONS_NEW);
        this.handlers = inject(TuiItemsHandlersDirective);
        this.dropdownEnabled = tuiDropdownEnabled(computed(() => !this.native && this.interactive()));
        this.options = inject(TUI_INPUT_DATE_OPTIONS_NEW);
        this.format = toSignal(inject(TUI_DATE_FORMAT), {
            initialValue: TUI_DEFAULT_DATE_FORMAT,
        });
        this.valueEffect = effect(() => {
            const value = this.stringify(this.value()) ||
                (this.filler().length === this.el.value.length ? '' : this.el.value);
            this.textfield.value.set(value);
        }, TUI_ALLOW_SIGNAL_WRITES);
        this.calendarIn = effect(() => {
            if (this.calendar()) {
                this.processCalendar(this.calendar());
            }
        }, TUI_ALLOW_SIGNAL_WRITES);
        this.calendarOut = effect((onCleanup) => {
            const subscription = this.calendar()?.valueChange.subscribe((value) => this.setDate(value));
            onCleanup(() => subscription?.unsubscribe());
        });
        this.native = !!inject(TuiWithNativePicker, { optional: true }) && this.mobile;
        this.min = signal(this.options.min);
        this.max = signal(this.options.max);
    }
    set minSetter(min) {
        this.min.set(min instanceof TuiDay ? min : this.options.min);
    }
    set maxSetter(max) {
        this.max.set(max instanceof TuiDay ? max : this.options.max);
    }
    writeValue(value) {
        const reset = this.control.pristine && this.control.untouched && !value;
        const changed = untracked(() => value !== this.value());
        if (changed || reset) {
            super.writeValue(value);
            this.textfield.value.set(this.stringify(this.value()));
        }
    }
    setDate(value) {
        this.onChange(value);
        this.open.set(false);
        if (!this.el.closest('tui-dropdown')) {
            this.el.blur();
        }
    }
    processCalendar(calendar) {
        const value = this.value();
        calendar.value = Array.isArray(value) ? value[0] : value;
        calendar.disabledItemHandler =
            this.handlers.disabledItemHandler();
        calendar.min = this.min();
        calendar.max = this.max();
    }
    onClick() {
        if (!this.mobile && this.dropdownEnabled()) {
            this.open.update((open) => !open);
        }
    }
    stringify(value) {
        return value?.toString(this.format().mode, this.format().separator) ?? '';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateBase, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiInputDateBase, isStandalone: true, inputs: { minSetter: ["min", "minSetter"], maxSetter: ["max", "maxSetter"] }, host: { listeners: { "input": "onValueChange($event.target.value)", "click.capture.stop": "onClick()" }, properties: { "attr.inputmode": "mobile && open() ? \"none\" : \"numeric\"", "disabled": "disabled()" } }, usesInheritance: true, ngImport: i0 }); }
}
export { TuiInputDateBase };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateBase, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    host: {
                        '[attr.inputmode]': 'mobile && open() ? "none" : "numeric"',
                        '[disabled]': 'disabled()',
                        '(input)': 'onValueChange($event.target.value)',
                        '(click.capture.stop)': 'onClick()',
                    },
                }]
        }], propDecorators: { minSetter: [{
                type: Input,
                args: ['min']
            }], maxSetter: [{
                type: Input,
                args: ['max']
            }] } });
class TuiInputDateDirective extends TuiInputDateBase {
    constructor() {
        super(...arguments);
        this.identity = inject(TUI_ITEMS_HANDLERS).identityMatcher.set((a, b) => a.daySame(b));
        this.mask = tuiMaskito(computed(() => maskitoDateOptionsGenerator({
            separator: this.format().separator,
            mode: TUI_DATE_ADAPTER[this.format().mode],
            min: this.min().toLocalNativeDate(),
            max: this.max().toLocalNativeDate(),
        })));
    }
    onValueChange(value) {
        this.control?.control?.updateValueAndValidity({ emitEvent: false });
        this.onChange(value.length === DATE_FILLER_LENGTH
            ? TuiDay.normalizeParse(value, this.format().mode)
            : null);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiInputDateDirective, isStandalone: true, selector: "input[tuiInputDate]", providers: [
            tuiAsOptionContent(TuiSelectOption),
            tuiAsControl(TuiInputDateDirective),
            tuiValueTransformerFrom(TUI_INPUT_DATE_OPTIONS_NEW),
        ], usesInheritance: true, hostDirectives: [{ directive: i1.TuiWithTextfield }, { directive: i2.TuiDropdownAuto }, { directive: i3.TuiItemsHandlersValidator }, { directive: i4.MaskitoDirective }], ngImport: i0 }); }
}
export { TuiInputDateDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'input[tuiInputDate]',
                    providers: [
                        tuiAsOptionContent(TuiSelectOption),
                        tuiAsControl(TuiInputDateDirective),
                        tuiValueTransformerFrom(TUI_INPUT_DATE_OPTIONS_NEW),
                    ],
                    hostDirectives: [
                        TuiWithTextfield,
                        TuiDropdownAuto,
                        TuiItemsHandlersValidator,
                        MaskitoDirective,
                    ],
                }]
        }] });
