// @taiga-ui/core/esm2022/directives/dropdown/dropdown-a11y.directive.mjs
import { Directive, effect, inject, Input, signal } from '@angular/core';
import { TuiIdService } from '@taiga-ui/cdk/services';
import { TuiDropdownDirective } from './dropdown.directive';
import { TUI_DROPDOWN_HOST } from './dropdown.providers';
import * as i0 from "@angular/core";
class TuiDropdownA11y {
    constructor() {
        this.id = inject(TuiIdService).generate();
        this.host = inject(TUI_DROPDOWN_HOST);
        this.dropdown = inject(TuiDropdownDirective);
        this.sync = effect(() => {
            const content = this.dropdown._content();
            const dropdown = this.dropdown.ref();
            const host = this.host.nativeElement;
            host.setAttribute('aria-expanded', String(!!dropdown));
            host.setAttribute('aria-controls', this.id);
            host.setAttribute('aria-haspopup', this._tuiDropdownRole());
            if (host.matches('input')) {
                if (content) {
                    host.setAttribute('role', 'combobox');
                }
                else {
                    host.removeAttribute('role');
                }
            }
            dropdown?.location.nativeElement.setAttribute('role', this._tuiDropdownRole());
            dropdown?.location.nativeElement.setAttribute('id', this.id);
            if (content) {
                return;
            }
            host.removeAttribute('aria-expanded');
            host.removeAttribute('aria-controls');
            host.removeAttribute('aria-haspopup');
        });
        // TODO: refactor to signal inputs after Angular update
        // eslint-disable-next-line @typescript-eslint/naming-convention
        this._tuiDropdownRole = signal('listbox');
    }
    set tuiDropdownRole(role) {
        this._tuiDropdownRole.set(role);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDropdownA11y, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiDropdownA11y, isStandalone: true, selector: "[tuiDropdownA11y]", inputs: { tuiDropdownRole: "tuiDropdownRole" }, ngImport: i0 }); }
}
export { TuiDropdownA11y };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDropdownA11y, decorators: [{
            type: Directive,
            args: [{ standalone: true, selector: '[tuiDropdownA11y]' }]
        }], propDecorators: { tuiDropdownRole: [{
                type: Input
            }] } });
