// @taiga-ui/core/esm2022/components/dialog/active-zone-adapter.directive.mjs
import { DOCUMENT } from '@angular/common';
import { Directive, inject } from '@angular/core';
import { TuiActiveZone } from '@taiga-ui/cdk/directives/active-zone';
import { tuiGetNativeFocused } from '@taiga-ui/cdk/utils/focus';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk/directives/active-zone";
class TuiActiveZoneAdapter {
    constructor() {
        this.current = inject(TuiActiveZone);
        this.parent = findActive(inject(TuiActiveZone, { skipSelf: true }), tuiGetNativeFocused(inject(DOCUMENT)));
    }
    ngOnInit() {
        this.current.tuiActiveZoneParentSetter = this.parent;
    }
    ngOnDestroy() {
        this.current.tuiActiveZoneParentSetter = null;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiActiveZoneAdapter, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiActiveZoneAdapter, isStandalone: true, selector: "[tuiActiveZoneAdapter]", hostDirectives: [{ directive: i1.TuiActiveZone }], ngImport: i0 }); }
}
export { TuiActiveZoneAdapter };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiActiveZoneAdapter, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiActiveZoneAdapter]',
                    hostDirectives: [TuiActiveZone],
                }]
        }] });
function findActive(zone, element) {
    if (!element || !zone.contains(element)) {
        return null;
    }
    const active = zone.children.find((child) => child.contains(element));
    return active ? findActive(active, element) : zone;
}
