// @taiga-ui/core/esm2022/components/dialog/dialog.tokens.mjs
import { InjectionToken } from '@angular/core';
import { tuiProvideOptions } from '@taiga-ui/cdk/utils/miscellaneous';
import { BehaviorSubject, EMPTY } from 'rxjs';
export const TUI_DIALOGS = new InjectionToken(ngDevMode ? 'TUI_DIALOGS' : '', {
    factory: () => new BehaviorSubject([]),
});
export const TUI_DIALOG_DEFAULT_OPTIONS = {
    appearance: '',
    size: 'm',
    required: false,
    closeable: true,
    dismissible: true,
    label: '',
    header: '',
    data: undefined,
};
/**
 * A stream to close dialogs
 * TODO: use router.events in v5
 */
export const TUI_DIALOGS_CLOSE = new InjectionToken(ngDevMode ? 'TUI_DIALOGS_CLOSE' : '', {
    factory: () => EMPTY,
});
/**
 * Default parameters for dialog component
 */
export const TUI_DIALOG_OPTIONS = new InjectionToken(ngDevMode ? 'TUI_DIALOG_OPTIONS' : '', {
    factory: () => TUI_DIALOG_DEFAULT_OPTIONS,
});
export function tuiDialogOptionsProvider(options) {
    return tuiProvideOptions(TUI_DIALOG_OPTIONS, options, TUI_DIALOG_DEFAULT_OPTIONS);
}
