// @taiga-ui/layout/fesm2022/taiga-ui-layout-tokens.mjs
import { InjectionToken } from '@angular/core';
import { tuiProvideOptions } from '@taiga-ui/cdk/utils/miscellaneous';
import { tuiExtractI18n } from '@taiga-ui/i18n/utils';

const LAYOUT_ICONS = {
    filter: '@tui.filter',
    grid: '@tui.layout-grid',
};
const TUI_LAYOUT_ICONS = new InjectionToken(ngDevMode ? 'TUI_LAYOUT_ICONS' : '', {
    factory: () => LAYOUT_ICONS,
});
function tuiLayoutIconsProvider(icons) {
    return tuiProvideOptions(TUI_LAYOUT_ICONS, icons, LAYOUT_ICONS);
}
/**
 * @deprecated use {@link TUI_LAYOUT_ICONS} instead
 */
const TUI_COMMON_ICONS = TUI_LAYOUT_ICONS;
/**
 * @deprecated use {@link tuiLayoutIconsProvider} instead
 */
const tuiCommonIconsProvider = tuiLayoutIconsProvider;

const TUI_INPUT_SEARCH = new InjectionToken(ngDevMode ? 'TUI_INPUT_SEARCH' : '', {
    factory: tuiExtractI18n('inputSearch'),
});

/**
 * Generated bundle index. Do not edit.
 */

export { TUI_COMMON_ICONS, TUI_INPUT_SEARCH, TUI_LAYOUT_ICONS, tuiCommonIconsProvider, tuiLayoutIconsProvider };
